#' Write code for loading libraries
#'
#' @param lib vector of libraries
#'
#' @rdname wrLib
#' @export wrLib
#'
wrLib <- function(lib) {
  oup = ""
  for(iLib in lib){
    oup = paste0(oup, "library(", iLib, ") \n")
  }
  glue::glue(paste0(oup, "\n"))
}

#' Write code for loading objects for server.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrSVload
#' @export wrSVload
#'
wrSVload <- function(prefix) {
  glue::glue('{prefix}conf = readRDS("{prefix}conf.rds")\n',
             '{prefix}def  = readRDS("{prefix}def.rds")\n',
             '{prefix}gene = readRDS("{prefix}gene.rds")\n',
             '{prefix}meta = readRDS("{prefix}meta.rds")\n',
             '  \n',
             'if({prefix}conf$extra_tabs[1]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}m_all.rds"))) {{ \n',
             '    {prefix}m_all = readRDS("{prefix}m_all.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    #{prefix}m_all = matrix(c("ERROR: data file \'{prefix}m_all.rds\' not found, cannot create data table!", ""), nrow=1, ncol=1) \n',
             '    {prefix}m_all = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             'if({prefix}conf$extra_tabs[2]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}m_t20.rds"))) {{ \n',
             '    {prefix}m_t20 = readRDS("{prefix}m_t20.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    #{prefix}m_t20 = matrix(c("ERROR: data file \'{prefix}m_t20.rds\' not found, cannot create data table!", ""), nrow=1, ncol=1) \n',
             '    {prefix}m_t20 = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             'if({prefix}conf$extra_tabs[3]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}de_genes.rds"))) {{ \n',
             '    {prefix}de_genes = readRDS("{prefix}de_genes.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    #{prefix}de_genes = matrix(c("ERROR: data file \'{prefix}de_genes.rds\' not found, cannot create data table!", ""), nrow=1, ncol=1) \n',
             '    {prefix}de_genes = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             'if({prefix}conf$extra_tabs[4]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}gene_ranks.rds"))) {{ \n',
             '    {prefix}gene_ranks = readRDS("{prefix}gene_ranks.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    {prefix}gene_ranks = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             'if({prefix}conf$extra_tabs[5]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}de_genes_ggvolc.rds"))) {{ \n',
             '    {prefix}de_genes_ggvolc = readRDS("{prefix}de_genes_ggvolc.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    {prefix}de_genes_ggvolc = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             'if({prefix}conf$extra_tabs[6]==TRUE) {{ \n',
             '  if(file.exists(paste0(getwd(), "/{prefix}gene_ont.rds"))) {{ \n',
             '    {prefix}gene_ont = readRDS("{prefix}gene_ont.rds") \n',
             '  }} \n',
             '  else {{ \n',
             '    {prefix}gene_ont = matrix(c("ERROR")) \n',
             '  }} \n',
             '}} \n',
             '\n\n\n\n')
}

#' Write code for fixed portion of server.R
#'
#' @rdname wrSVfix
#' @export wrSVfix
#'
wrSVfix <- function() {
  glue::glue('### Useful stuff \n',
             '# Colour palette \n',
             'cList = list(c("grey85","#FFF7EC","#FEE8C8","#FDD49E","#FDBB84", \n',
             '               "#FC8D59","#EF6548","#D7301F","#B30000","#7F0000"), \n',
             '             c("#4575B4","#74ADD1","#ABD9E9","#E0F3F8","#FFFFBF", \n',
             '               "#FEE090","#FDAE61","#F46D43","#D73027")[c(1,1:9,9)], \n',
             '             c("#FDE725","#AADC32","#5DC863","#27AD81","#21908C", \n',
             '               "#2C728E","#3B528B","#472D7B","#440154")) \n',
             'names(cList) = c("White-Red", "Blue-Yellow-Red", "Yellow-Green-Purple") \n',
             ' \n',
             '# Panel sizes \n',
             'pList = c("400px", "600px", "800px") \n',
             'names(pList) = c("Small", "Medium", "Large") \n',
             'pList2 = c("500px", "700px", "900px") \n',
             'names(pList2) = c("Small", "Medium", "Large") \n',
             'pList3 = c("600px", "800px", "1000px") \n',
             'names(pList3) = c("Small", "Medium", "Large") \n',
             'sList = c(18,24,30) \n',
             'names(sList) = c("Small", "Medium", "Large") \n',
             'lList = c(5,6,7) \n',
             'names(lList) = c("Small", "Medium", "Large") \n',
             'sList2 = c(10,12,16) \n',
             'names(sList2) = c("Small", "Medium", "Large") \n',
             ' \n',
             '# Function to extract legend \n',
             'g_legend <- function(a.gplot){{  \n',
             '  tmp <- ggplot_gtable(ggplot_build(a.gplot))  \n',
             '  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")  \n',
             '  legend <- tmp$grobs[[leg]]  \n',
             '  legend \n',
             '}}  \n',
             ' \n',
             '# Plot theme \n',
             'sctheme <- function(base_size = 24, XYval = TRUE, Xang = 0, XjusH = 0.5){{ \n',
             '  oupTheme = theme( \n',
             '    text =             element_text(size = base_size, family = "Helvetica"), \n',
             '    panel.background = element_rect(fill = "white", colour = NA), \n',
             '    axis.line =   element_line(colour = "black"), \n',
             '    axis.ticks =  element_line(colour = "black", size = base_size / 20), \n',
             '    axis.title =  element_text(face = "bold"), \n',
             '    axis.text =   element_text(size = base_size), \n',
             '    axis.text.x = element_text(angle = Xang, hjust = XjusH), \n',
             '    legend.position = "bottom", \n',
             '    legend.key =      element_rect(colour = NA, fill = NA) \n',
             '  ) \n',
             '  if(!XYval){{ \n',
             '    oupTheme = oupTheme + theme( \n',
             '      axis.text.x = element_blank(), axis.ticks.x = element_blank(), \n',
             '      axis.text.y = element_blank(), axis.ticks.y = element_blank()) \n',
             '  }} \n',
             '  return(oupTheme) \n',
             '}} \n',
             ' \n',
             '### Common plotting functions \n',
             '# Plot cell information on dimred \n',
             'scDRcell <- function(inpConf, inpMeta, inpdrX, inpdrY, inp1, inpsub1, inpsub2, \n',
             '                     inpsiz, inpcol, inpord, inpfsz, inpasp, inptxt, inplab, inplegend=TRUE, insplit = NULL, split_idx = NULL, select = NULL){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  if (!(is.null(insplit))) {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                          inpConf[UI == inp1]$ID, inpConf[UI == inpsub1]$ID, \n',
             '                          inpConf[UI == insplit]$ID), with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "val", "sub", "split") \n',
             '    split_options = as.character(unique(inpMeta[[insplit]])) \n',
             '    #TODO: simplify this to only need the argument \'select\' for either condition \n',
             '      if(!is.null(select)) {{ \n',
             '        ggData = ggData[ggData$split == select] \n',
             '      }} \n',
             '      else {{ \n',
             '        ggData = ggData[ggData$split == split_options[split_idx]] \n',
             '      }} \n',            
             '  }} else {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                       inpConf[UI == inp1]$ID, inpConf[UI == inpsub1]$ID),  \n',
             '                        with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "val", "sub") \n',
             '  }} \n',
             '  rat = (max(ggData$X) - min(ggData$X)) / (max(ggData$Y) - min(ggData$Y)) \n',
             '  bgCells = FALSE \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    bgCells = TRUE \n',
             '    ggData2 = ggData[!sub %in% inpsub2] \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  if(inpord == "Max-1st"){{ \n',
             '    ggData = ggData[order(val)] \n',
             '  }} else if(inpord == "Min-1st"){{ \n',
             '    ggData = ggData[order(-val)] \n',
             '  }} else if(inpord == "Random"){{ \n',
             '    ggData = ggData[sample(nrow(ggData))] \n',
             '  }} \n',
             '  \n',
             '  # Do factoring if required \n',
             '  if(!is.na(inpConf[UI == inp1]$fCL)){{ \n',
             '    ggCol = strsplit(inpConf[UI == inp1]$fCL, "\\\\|")[[1]] \n',
             '    names(ggCol) = levels(ggData$val) \n',
             '    ggLvl = levels(ggData$val)[levels(ggData$val) %in% unique(ggData$val)] \n',
             '    ggData$val = factor(ggData$val, levels = ggLvl) \n',
             '    ggCol = ggCol[ggLvl] \n',
             '  }} \n',
             ' \n',
             '  # Actual ggplot \n',
             '  ggOut = ggplot(ggData, aes(X, Y, color = val)) \n',
             '  if(bgCells){{ \n',
             '    ggOut = ggOut + \n',
             '      geom_point(data = ggData2, color = "snow2", size = inpsiz, shape = 16) \n',
             '  }} \n',
             '  ggOut = ggOut + \n',
             '    geom_point(size = inpsiz, shape = 16) + xlab(inpdrX) + ylab(inpdrY) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = inptxt) \n',
             '  if(is.na(inpConf[UI == inp1]$fCL)){{ \n',
             '    ggOut = ggOut + scale_color_gradientn("", colours = cList[[inpcol]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) \n',
             '  }} else {{ \n',
             '    sListX = min(nchar(paste0(levels(ggData$val), collapse = "")), 200) \n',
             '    sListX = 0.75 * (sList - (1.5 * floor(sListX/50))) \n',
             '    ggOut = ggOut + scale_color_manual("", values = ggCol) + \n',
             '      guides(color = guide_legend(override.aes = list(size = 5),  \n',
             '                                  nrow = inpConf[UI == inp1]$fRow)) + \n',
             '      theme(legend.text = element_text(size = sListX[inpfsz])) \n',
             '    if(inplab){{ \n',
             '      ggData3 = ggData[, .(X = mean(X), Y = mean(Y)), by = "val"] \n',
             '      lListX = min(nchar(paste0(ggData3$val, collapse = "")), 200) \n',
             '      lListX = lList - (0.25 * floor(lListX/50)) \n',
             '      ggOut = ggOut + \n',
             '        geom_text_repel(data = ggData3, aes(X, Y, label = val), \n',
             '                        color = "grey10", bg.color = "grey95", bg.r = 0.15, \n',
             '                        size = lListX[inpfsz], seed = 42) \n',
             '    }} \n',
             '  }} \n',
             '  if(inpasp == "Square") {{ \n',
             '    ggOut = ggOut + coord_fixed(ratio = rat) \n',
             '  }} else if(inpasp == "Fixed") {{ \n',
             '    ggOut = ggOut + coord_fixed() \n',
             '  }} \n',
             '  #TODO: simplify this to only need the argument \'select\' for either condition \n',
             '  if (!(is.null(insplit))){{ \n',
             '    ggOut = ggOut + ggtitle(split_options[split_idx]) + theme(plot.title = element_text(hjust=0.5)) \n',
             '  }} \n',
             '  if(!(is.null(select))) {{ \n',
             '    ggOut = ggOut + ggtitle(select) + theme(plot.title = element_text(hjust=0.5)) \n',
             '  }} \n',
             '  if(inplegend == FALSE) {{ \n',
             '    ggOut = ggOut + theme(legend.position = "none") \n',
             '  }} \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             'scDRnum <- function(inpConf, inpMeta, inp1, inp2, inpsub1, inpsub2, \n',
             '                    inpH5, inpGene, inpsplt){{ \n', 
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  ggData = inpMeta[, c(inpConf[UI == inp1]$ID, inpConf[UI == inpsub1]$ID), \n',
             '                   with = FALSE] \n',
             '  colnames(ggData) = c("group", "sub") \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData$val2 = h5data$read(args = list(inpGene[inp2], quote(expr=))) \n',
             '  ggData[val2 < 0]$val2 = 0 \n',
             '  h5file$close_all() \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n', 
             '  \n',
             '  # Split inp1 if necessary \n',
             '  if(is.na(inpConf[UI == inp1]$fCL)){{ \n',
             '    if(inpsplt == "Quartile"){{nBk = 4}} \n', 
             '    if(inpsplt == "Decile"){{nBk = 10}} \n', 
             '    ggData$group = cut(ggData$group, breaks = nBk) \n',
             '  }} \n',
             '  \n',
             '  # Actual data.table \n',
             '  ggData$express = FALSE \n',
             '  ggData[val2 > 0]$express = TRUE \n',
             '  ggData1 = ggData[express == TRUE, .(nExpress = .N), by = "group"] \n',
             '  ggData = ggData[, .(nCells = .N), by = "group"] \n',
             '  ggData = ggData1[ggData, on = "group"] \n',
             '  ggData = ggData[, c("group", "nCells", "nExpress"), with = FALSE] \n',
             '  ggData[is.na(nExpress)]$nExpress = 0 \n',
             '  ggData$pctExpress = 100 * ggData$nExpress / ggData$nCells \n',
             '  ggData = ggData[order(group)] \n',
             '  colnames(ggData)[3] = paste0(colnames(ggData)[3], "_", inp2) \n',
             '  return(ggData) \n',
             '}} \n',
             ' \n',
             '# Plot gene expression on dimred \n',
             'scDRgene <- function(inpConf, inpMeta, inpdrX, inpdrY, inp1, inpsub1, inpsub2, \n',
             '                     inpH5, inpGene, \n',
             '                     inpsiz, inpcol, inpord, inpfsz, inpasp, inptxt, insplit = NULL, split_idx = NULL, select = NULL){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  if (!(is.null(insplit))){{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '    inpConf[UI == inpsub1]$ID, inpConf[UI == insplit]$ID), with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "sub", "split") \n',
             '    ggData$val = h5data$read(args = list(inpGene[inp1], quote(expr=))) \n',
             '    ggData[val < 0]$val = 0 \n',
             '    split_options = as.character(unique(inpMeta[[insplit]])) \n',
             '    #TODO: simplify this to only need the argument \'select\' for either condition \n',
             '    if(!is.null(select)) {{ \n',
             '      ggData = ggData[ggData$split == select] \n',
             '    }} \n', 
             '    else {{ \n',
             '      ggData = ggData[ggData$split == split_options[split_idx]] \n',
             '    }} \n', 
             '  }} else {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '    inpConf[UI == inpsub1]$ID), \n',
             '    with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "sub") \n',
             ' \n',
             '    ggData$val = h5data$read(args = list(inpGene[inp1], quote(expr=))) \n',
             '    ggData[val < 0]$val = 0 \n',
             '  }} \n',
             ' \n',
             '  h5file$close_all() \n', 
             '  rat = (max(ggData$X) - min(ggData$X)) / (max(ggData$Y) - min(ggData$Y)) \n',
             '  \n',
             '  bgCells = FALSE \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    bgCells = TRUE \n',
             '    ggData2 = ggData[!sub %in% inpsub2] \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  if(inpord == "Max-1st"){{ \n',
             '    ggData = ggData[order(val)] \n',
             '  }} else if(inpord == "Min-1st"){{ \n',
             '    ggData = ggData[order(-val)] \n',
             '  }} else if(inpord == "Random"){{ \n',
             '    ggData = ggData[sample(nrow(ggData))] \n',
             '  }} \n',
             '   \n',
             '  # Actual ggplot \n',
             '  ggOut = ggplot(ggData, aes(X, Y, color = val)) \n',
             '  if(bgCells){{ \n',
             '    ggOut = ggOut + \n',
             '      geom_point(data = ggData2, color = "snow2", size = inpsiz, shape = 16) \n',
             '  }} \n',
             '  ggOut = ggOut + \n',
             '    geom_point(size = inpsiz, shape = 16) + xlab(inpdrX) + ylab(inpdrY) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = inptxt) +  \n',
             '    scale_color_gradientn(inp1, colours = cList[[inpcol]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) \n',
             '  if(inpasp == "Square") {{ \n',
             '    ggOut = ggOut + coord_fixed(ratio = rat) \n',
             '  }} else if(inpasp == "Fixed") {{ \n',
             '    ggOut = ggOut + coord_fixed() \n',
             '  }} \n',
             '  if (!(is.null(insplit))) {{ \n',
             '    ggOut = ggOut + ggtitle(split_options[split_idx]) + theme(plot.title = element_text(hjust=0.5)) \n',
             '  }} \n',
             '  if(!(is.null(select))) {{ \n',
             '    ggOut = ggOut + ggtitle(select) + theme(plot.title = element_text(hjust=0.5)) \n',
             '  }} \n', 
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             '# Plot gene coexpression on dimred \n',
             'bilinear <- function(x,y,xy,Q11,Q21,Q12,Q22){{ \n',
             '  oup = (xy-x)*(xy-y)*Q11 + x*(xy-y)*Q21 + (xy-x)*y*Q12 + x*y*Q22 \n',
             '  oup = oup / (xy*xy) \n',
             '  return(oup) \n',
             '}} \n',
             ' \n',
             'scDRcoex <- function(inpConf, inpMeta, inpdrX, inpdrY, inp1, inp2, \n',
             '                     inpsub1, inpsub2, inpH5, inpGene, \n',
             '                     inpsiz, inpcol, inpord, inpfsz, inpasp, inptxt, inpSplit=NULL, inpSplitParam=NULL){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  if(is.null(inpSplit)) {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                         inpConf[UI == inpsub1]$ID),  \n',
             '                     with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "sub") \n',
             '  }} \n',
             '  else {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpdrX]$ID, inpConf[UI == inpdrY]$ID, \n',
             '                         inpConf[UI == inpsub1]$ID,  inpConf[UI == inpSplit]$ID), \n',
             '                     with = FALSE] \n',
             '    colnames(ggData) = c("X", "Y", "sub", inpSplit) \n',
             '  }} \n',
             '  rat = (max(ggData$X) - min(ggData$X)) / (max(ggData$Y) - min(ggData$Y)) \n',
             '  \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData$val1 = h5data$read(args = list(inpGene[inp1], quote(expr=))) \n',
             '  ggData[val1 < 0]$val1 = 0 \n',
             '  ggData$val2 = h5data$read(args = list(inpGene[inp2], quote(expr=))) \n',
             '  ggData[val2 < 0]$val2 = 0 \n',
             '  h5file$close_all() \n',
             '  bgCells = FALSE \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    bgCells = TRUE \n',
             '    ggData2 = ggData[!sub %in% inpsub2] \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  \n',
             '  # Generate coex color palette \n',
             '  cInp = strsplit(inpcol, "; ")[[1]] \n',
             '  if(cInp[1] == "Red (Gene1)"){{ \n',
             '    c10 = c(255,0,0) \n',
             '  }} else if(cInp[1] == "Orange (Gene1)"){{ \n',
             '    c10 = c(255,140,0) \n',
             '  }} else {{ \n',
             '    c10 = c(0,255,0) \n',
             '  }} \n',
             '  if(cInp[2] == "Green (Gene2)"){{ \n',
             '    c01 = c(0,255,0) \n',
             '  }} else {{ \n',
             '    c01 = c(0,0,255) \n',
             '  }} \n',
             '  c00 = c(217,217,217) ; c11 = c10 + c01 \n',
             '  nGrid = 16; nPad = 2; nTot = nGrid + nPad * 2 \n',
             '  gg = data.table(v1 = rep(0:nTot,nTot+1), v2 = sort(rep(0:nTot,nTot+1))) \n',
             '  gg$vv1 = gg$v1 - nPad ; gg[vv1 < 0]$vv1 = 0; gg[vv1 > nGrid]$vv1 = nGrid \n',
             '  gg$vv2 = gg$v2 - nPad ; gg[vv2 < 0]$vv2 = 0; gg[vv2 > nGrid]$vv2 = nGrid \n',
             '  gg$cR = bilinear(gg$vv1, gg$vv2, nGrid, c00[1], c10[1], c01[1], c11[1]) \n',
             '  gg$cG = bilinear(gg$vv1, gg$vv2, nGrid, c00[2], c10[2], c01[2], c11[2]) \n',
             '  gg$cB = bilinear(gg$vv1, gg$vv2, nGrid, c00[3], c10[3], c01[3], c11[3]) \n',
             '  gg$cMix = rgb(gg$cR, gg$cG, gg$cB, maxColorValue = 255) \n',
             '  gg = gg[, c("v1", "v2", "cMix")] \n',
             '  \n',
             '  # Map colours \n',
             '  ggData$v1 = round(nTot * ggData$val1 / max(ggData$val1)) \n',
             '  ggData$v2 = round(nTot * ggData$val2 / max(ggData$val2)) \n',
             '  ggData$v0 = ggData$v1 + ggData$v2 \n',
             '  ggData = gg[ggData, on = c("v1", "v2")] \n',
             '  if(inpord == "Max-1st"){{ \n',
             '    ggData = ggData[order(v0)] \n',
             '  }} else if(inpord == "Min-1st"){{ \n',
             '    ggData = ggData[order(-v0)] \n',
             '  }} else if(inpord == "Random"){{ \n',
             '    ggData = ggData[sample(nrow(ggData))] \n',
             '  }} \n',
             '  \n',
             '  # Actual ggplot \n',
             '  if(!is.null(inpSplit)) {{ \n',
             '    ggData <- subset(ggData, ggData[[inpSplit]] == inpSplitParam) \n',
             '    ggOut = ggplot(ggData, aes(X, Y)) + labs(title = inpSplitParam) \n',
             '  }} \n',
             '  else {{ \n',    
             '    ggOut = ggplot(ggData, aes(X, Y)) \n',
             '  }} \n',
             '  if(bgCells){{ \n',
             '    ggOut = ggOut + \n',
             '      geom_point(data = ggData2, color = "snow2", size = inpsiz, shape = 16) \n',
             '  }} \n',
             '  ggOut = ggOut + \n',
             '    geom_point(size = inpsiz, shape = 16, color = ggData$cMix) + \n',
             '    xlab(inpdrX) + ylab(inpdrY) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = inptxt) + \n',
             '    scale_color_gradientn(inp1, colours = cList[[1]]) + \n',
             '    guides(color = guide_colorbar(barwidth = 15)) \n',
             '  if(inpasp == "Square") {{ \n',
             '    ggOut = ggOut + coord_fixed(ratio = rat) \n',
             '  }} else if(inpasp == "Fixed") {{ \n',
             '    ggOut = ggOut + coord_fixed() \n',
             '  }} \n',
             '  \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             'scDRcoexLeg <- function(inp1, inp2, inpcol, inpfsz){{ \n',
             '  # Generate coex color palette \n',
             '  cInp = strsplit(inpcol, "; ")[[1]] \n',
             '  if(cInp[1] == "Red (Gene1)"){{ \n',
             '    c10 = c(255,0,0) \n',
             '  }} else if(cInp[1] == "Orange (Gene1)"){{ \n',
             '    c10 = c(255,140,0) \n',
             '  }} else {{ \n',
             '    c10 = c(0,255,0) \n',
             '  }} \n',
             '  if(cInp[2] == "Green (Gene2)"){{ \n',
             '    c01 = c(0,255,0) \n',
             '  }} else {{ \n',
             '    c01 = c(0,0,255) \n',
             '  }} \n',
             '  c00 = c(217,217,217) ; c11 = c10 + c01 \n',
             '  nGrid = 16; nPad = 2; nTot = nGrid + nPad * 2 \n',
             '  gg = data.table(v1 = rep(0:nTot,nTot+1), v2 = sort(rep(0:nTot,nTot+1))) \n',
             '  gg$vv1 = gg$v1 - nPad ; gg[vv1 < 0]$vv1 = 0; gg[vv1 > nGrid]$vv1 = nGrid \n',
             '  gg$vv2 = gg$v2 - nPad ; gg[vv2 < 0]$vv2 = 0; gg[vv2 > nGrid]$vv2 = nGrid \n',
             '  gg$cR = bilinear(gg$vv1, gg$vv2, nGrid, c00[1], c10[1], c01[1], c11[1]) \n',
             '  gg$cG = bilinear(gg$vv1, gg$vv2, nGrid, c00[2], c10[2], c01[2], c11[2]) \n',
             '  gg$cB = bilinear(gg$vv1, gg$vv2, nGrid, c00[3], c10[3], c01[3], c11[3]) \n',
             '  gg$cMix = rgb(gg$cR, gg$cG, gg$cB, maxColorValue = 255) \n',
             '  gg = gg[, c("v1", "v2", "cMix")] \n',
             '  \n',
             '  # Actual ggplot \n',
             '  ggOut = ggplot(gg, aes(v1, v2)) + \n',
             '    geom_tile(fill = gg$cMix) + \n',
             '    xlab(inp1) + ylab(inp2) + coord_fixed(ratio = 1) + \n',
             '    scale_x_continuous(breaks = c(0, nTot), label = c("low", "high")) + \n',
             '    scale_y_continuous(breaks = c(0, nTot), label = c("low", "high")) + \n',
             '    sctheme(base_size = sList[inpfsz], XYval = TRUE) \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             'scDRcoexNum <- function(inpConf, inpMeta, inp1, inp2, \n',
             '                        inpsub1, inpsub2, inpH5, inpGene, inpSplit=NULL, inpSplitParam=NULL){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  if(is.null(inpSplit)) {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpsub1]$ID), with = FALSE] \n',
             '    colnames(ggData) = c("sub") \n',
             '  }} \n',
             '  else {{ \n',
             '    ggData = inpMeta[, c(inpConf[UI == inpsub1]$ID, inpConf[UI == inpSplit]$ID), with = FALSE] \n',
             '    colnames(ggData) = c("sub", inpSplit) \n',
             '  }} \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData$val1 = h5data$read(args = list(inpGene[inp1], quote(expr=))) \n',
             '  ggData[val1 < 0]$val1 = 0 \n',
             '  ggData$val2 = h5data$read(args = list(inpGene[inp2], quote(expr=))) \n',
             '  ggData[val2 < 0]$val2 = 0 \n',
             '  h5file$close_all() \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  if(!is.null(inpSplit)) {{ \n',
             '    ggData <- subset(ggData, ggData[[inpSplit]] == inpSplitParam) \n',
             '  }} \n',
             '  \n',
             '  # Actual data.table \n',
             '  ggData$express = "none" \n',
             '  ggData[val1 > 0]$express = inp1 \n',
             '  ggData[val2 > 0]$express = inp2 \n',
             '  ggData[val1 > 0 & val2 > 0]$express = "both" \n',
             '  ggData$express = factor(ggData$express, levels = unique(c("both", inp1, inp2, "none"))) \n',
             '  ggData = ggData[, .(nCells = .N), by = "express"] \n',
             '  ggData$percent = 100 * ggData$nCells / sum(ggData$nCells) \n',
             '  ggData = ggData[order(express)] \n',
             '  colnames(ggData)[1] = "expression > 0" \n',
             '  return(ggData) \n',
             '}} \n',
             ' \n',
             '# Plot violin / boxplot \n',
             'scVioBox <- function(inpConf, inpMeta, inp1, inp2, \n',
             '                     inpsub1, inpsub2, inpH5, inpGene, \n',     
             '                     inptyp, inppts, inpsiz, inpfsz){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  ggData = inpMeta[, c(inpConf[UI == inp1]$ID, inpConf[UI == inpsub1]$ID), \n',
             '                   with = FALSE] \n',
             '  colnames(ggData) = c("X", "sub") \n',
             '  \n',
             '  # Load in either cell meta or gene expr\n',
             '  if(inp2 %in% inpConf$UI){{ \n',
             '    ggData$val = inpMeta[[inpConf[UI == inp2]$ID]] \n',
             '  }} else {{ \n',
             '    h5file <- H5File$new(inpH5, mode = "r") \n',
             '    h5data <- h5file[["grp"]][["data"]] \n',
             '    ggData$val = h5data$read(args = list(inpGene[inp2], quote(expr=))) \n',
             '    ggData[val < 0]$val = 0 \n',
             '    set.seed(42) \n',
             '    tmpNoise = rnorm(length(ggData$val)) * diff(range(ggData$val)) / 1000 \n',
             '    ggData$val = ggData$val + tmpNoise \n',
             '    h5file$close_all() \n',
             '  }} \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  \n',
             '  # Do factoring \n',
             '  ggCol = strsplit(inpConf[UI == inp1]$fCL, "\\\\|")[[1]] \n',
             '  names(ggCol) = levels(ggData$X) \n',
             '  ggLvl = levels(ggData$X)[levels(ggData$X) %in% unique(ggData$X)] \n',
             '  ggData$X = factor(ggData$X, levels = ggLvl) \n',
             '  ggCol = ggCol[ggLvl] \n',
             '  \n',
             '  # Actual ggplot \n',
             '  if(inptyp == "violin"){{ \n',
             '    ggOut = ggplot(ggData, aes(X, val, fill = X)) + geom_violin(scale = "width") \n',
             '  }} else {{ \n',
             '    ggOut = ggplot(ggData, aes(X, val, fill = X)) + geom_boxplot() \n',
             '  }} \n',
             '  if(inppts){{ \n',
             '    ggOut = ggOut + geom_jitter(size = inpsiz, shape = 16) \n',
             '  }} \n',
             '  ggOut = ggOut + xlab(inp1) + ylab(inp2) + \n',
             '    sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) +  \n',
             '    scale_fill_manual("", values = ggCol) +\n',
             '    theme(legend.position = "none")\n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             '# Plot proportion plot \n',
             'scProp <- function(inpConf, inpMeta, inp1, inp2, inpsub1, inpsub2, \n',
             '                   inptyp, inpflp, inpfsz){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Prepare ggData \n',
             '  ggData = inpMeta[, c(inpConf[UI == inp1]$ID, inpConf[UI == inp2]$ID, \n',
             '                       inpConf[UI == inpsub1]$ID),  \n',
             '                   with = FALSE] \n',
             '  colnames(ggData) = c("X", "grp", "sub") \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  ggData = ggData[, .(nCells = .N), by = c("X", "grp")] \n',
             '  ggData = ggData[, {{tot = sum(nCells) \n',
             '                      .SD[,.(pctCells = 100 * sum(nCells) / tot, \n',
             '                             nCells = nCells), by = "grp"]}}, by = "X"] \n',
             '  \n',
             '  # Do factoring \n',
             '  ggCol = strsplit(inpConf[UI == inp2]$fCL, "\\\\|")[[1]] \n',
             '  names(ggCol) = levels(ggData$grp) \n',
             '  ggLvl = levels(ggData$grp)[levels(ggData$grp) %in% unique(ggData$grp)] \n',
             '  ggData$grp = factor(ggData$grp, levels = ggLvl) \n',
             '  ggCol = ggCol[ggLvl] \n',
             '  \n',
             '  # Actual ggplot \n',
             '  if(inptyp == "Proportion"){{ \n',
             '    ggOut = ggplot(ggData, aes(X, pctCells, fill = grp)) + \n',
             '      geom_col() + ylab("Cell Proportion (%)") \n',
             '  }} else {{ \n',
             '    ggOut = ggplot(ggData, aes(X, nCells, fill = grp)) + \n',
             '      geom_col() + ylab("Number of Cells") \n',
             '  }} \n',
             '  if(inpflp){{ \n',
             '    ggOut = ggOut + coord_flip() \n',
             '  }} \n',
             '  ggOut = ggOut + xlab(inp1) + \n',
             '    sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) +  \n',
             '    scale_fill_manual("", values = ggCol) + \n',
             '    theme(legend.position = "right") \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             '# Get gene list \n',
             'scGeneList <- function(inp, inpGene){{ \n',
             '  geneList = data.table(gene = unique(trimws(strsplit(inp, ",|;|\n")[[1]])), \n',
             '                        present = TRUE) \n',
             '  geneList[!gene %in% names(inpGene)]$present = FALSE \n',
             '  return(geneList) \n',
             '}} \n',
             ' \n',
             '# Plot gene expression bubbleplot / heatmap \n',
             'scBubbHeat <- function(inpConf, inpMeta, inp, inpGrp, inpPlt, \n',
             '                       inpsub1, inpsub2, inpH5, inpGene, inpScl, inpRow, inpCol, \n',
             '                       inpcols, inpfsz, save = FALSE){{ \n',
             '  if(is.null(inpsub1)){{inpsub1 = inpConf$UI[1]}} \n',
             '  # Identify genes that are in our dataset \n',
             '  geneList = scGeneList(inp, inpGene) \n',
             '  geneList = geneList[present == TRUE] \n',
             '  shiny::validate(need(nrow(geneList) <= 50, "More than 50 genes to plot! Please reduce the gene list!")) \n',
             '  shiny::validate(need(nrow(geneList) > 1, "Please input at least 2 genes to plot!")) \n',
             '   \n',
             '  # Prepare ggData \n',
             '  h5file <- H5File$new(inpH5, mode = "r") \n',
             '  h5data <- h5file[["grp"]][["data"]] \n',
             '  ggData = data.table() \n',
             '  for(iGene in geneList$gene){{ \n',
             '    tmp = inpMeta[, c("sampleID", inpConf[UI == inpsub1]$ID), with = FALSE] \n',
             '    colnames(tmp) = c("sampleID", "sub") \n',
             '    tmp$grpBy = inpMeta[[inpConf[UI == inpGrp]$ID]] \n',
             '    tmp$geneName = iGene \n',
             '    tmp$val = h5data$read(args = list(inpGene[iGene], quote(expr=))) \n',
             '    ggData = rbindlist(list(ggData, tmp)) \n',
             '  }} \n',
             '  h5file$close_all() \n',
             '  if(length(inpsub2) != 0 & length(inpsub2) != nlevels(ggData$sub)){{ \n',
             '    ggData = ggData[sub %in% inpsub2] \n',
             '  }} \n',
             '  shiny::validate(need(uniqueN(ggData$grpBy) > 1, "Only 1 group present, unable to plot!")) \n',
             '   \n',
             '  # Aggregate \n',
             '  #ggData$val = expm1(ggData$val) \n',
             '  ggData$val = scale(ggData$val) \n',
             '  ggData = ggData[, .(val = mean(val), prop = sum(val>0) / length(sampleID)), \n',
             '                  by = c("geneName", "grpBy")] \n',
             '  #ggData$val = log1p(ggData$val) \n',
             '   \n',
             '  # Scale if required \n',
             '  colRange = range(ggData$val) \n',
             '  if(inpScl){{ \n',
             '    ggData[, val:= scale(val), keyby = "geneName"] \n',
             '    colRange = c(-max(abs(range(ggData$val))), max(abs(range(ggData$val)))) \n',
             '  }} \n',
             '   \n',
             '  # hclust row/col if necessary \n',
             '  ggMat = dcast.data.table(ggData, geneName~grpBy, value.var = "val") \n',
             '  tmp = ggMat$geneName \n',
             '  ggMat = as.matrix(ggMat[, -1]) \n',
             '  rownames(ggMat) = tmp \n',
             '  if(inpRow){{ \n',
             '    hcRow = dendro_data(as.dendrogram(hclust(dist(ggMat)))) \n',
             '    ggRow = ggplot() + coord_flip() + \n',
             '      geom_segment(data = hcRow$segments, aes(x=x,y=y,xend=xend,yend=yend)) + \n',
             '      scale_y_continuous(breaks = rep(0, uniqueN(ggData$grpBy)), \n',
             '                         labels = unique(ggData$grpBy), expand = c(0, 0)) + \n',
             '      scale_x_continuous(breaks = seq_along(hcRow$labels$label), \n',
             '                         labels = hcRow$labels$label, expand = c(0, 0.5)) + \n',
             '      sctheme(base_size = sList[inpfsz]) + \n',
             '      theme(axis.title = element_blank(), axis.line = element_blank(), \n',
             '            axis.ticks = element_blank(), axis.text.y = element_blank(), \n',
             '            axis.text.x = element_text(color="white", angle = 45, hjust = 1)) \n',
             '    ggData$geneName = factor(ggData$geneName, levels = hcRow$labels$label) \n',
             '  }} else {{ \n',
             '    ggData$geneName = factor(ggData$geneName, levels = rev(geneList$gene)) \n',
             '  }} \n',
             '  if(inpCol){{ \n',
             '    hcCol = dendro_data(as.dendrogram(hclust(dist(t(ggMat))))) \n',
             '    ggCol = ggplot() + \n',
             '      geom_segment(data = hcCol$segments, aes(x=x,y=y,xend=xend,yend=yend)) + \n',
             '      scale_x_continuous(breaks = seq_along(hcCol$labels$label), \n',
             '                         labels = hcCol$labels$label, expand = c(0.05, 0)) + \n',
             '      scale_y_continuous(breaks = rep(0, uniqueN(ggData$geneName)), \n',
             '                         labels = unique(ggData$geneName), expand=c(0,0)) + \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) + \n',
             '      theme(axis.title = element_blank(), axis.line = element_blank(), \n',
             '            axis.ticks = element_blank(), axis.text.x = element_blank(), \n',
             '            axis.text.y = element_text(color = "white")) \n',
             '    ggData$grpBy = factor(ggData$grpBy, levels = hcCol$labels$label) \n',
             '  }} \n',
             '   \n',
             '  # Actual plot according to plottype \n',
             '  if(inpPlt == "Bubbleplot"){{ \n',
             '    # Bubbleplot \n',
             '    ggOut = ggplot(ggData, aes(grpBy, geneName, color = val, size = prop)) + \n',
             '      geom_point() +  \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) +  \n',
             '      scale_x_discrete(expand = c(0.05, 0)) +  \n',
             '      scale_y_discrete(expand = c(0, 0.5)) + \n',
             '      scale_size_continuous("proportion", range = c(0, 8), \n',
             '                            limits = c(0, 1), breaks = c(0.00,0.25,0.50,0.75,1.00)) + \n',
             '      scale_color_gradientn("expression", limits = colRange, colours = cList[[inpcols]]) + \n',
             '      guides(color = guide_colorbar(barwidth = 15)) + \n',
             '      theme(axis.title = element_blank(), legend.box = "vertical") \n',
             '  }} else {{ \n',
             '    # Heatmap \n',
             '    ggOut = ggplot(ggData, aes(grpBy, geneName, fill = val)) + \n',
             '      geom_tile() +  \n',
             '      sctheme(base_size = sList[inpfsz], Xang = 45, XjusH = 1) + \n',
             '      scale_x_discrete(expand = c(0.05, 0)) +  \n',
             '      scale_y_discrete(expand = c(0, 0.5)) + \n',
             '      scale_fill_gradientn("expression", limits = colRange, colours = cList[[inpcols]]) + \n',
             '      guides(fill = guide_colorbar(barwidth = 15)) + \n',
             '      theme(axis.title = element_blank()) \n',
             '  }} \n',
             '     \n',
             '  # Final tidy \n',
             '  ggLeg = g_legend(ggOut) \n',
             '  ggOut = ggOut + theme(legend.position = "none") \n',
             '  if(!save){{ \n',
             '    if(inpRow & inpCol){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggCol, ggRow, widths = c(7,1), heights = c(1,7,2),  \n',
             '                   layout_matrix = rbind(c(3,NA),c(1,4),c(2,NA)))  \n',
             '    }} else if(inpRow){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggRow, widths = c(7,1), heights = c(7,2),  \n',
             '                   layout_matrix = rbind(c(1,3),c(2,NA)))  \n',
             '    }} else if(inpCol){{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, ggCol, heights = c(1,7,2),  \n',
             '                   layout_matrix = rbind(c(3),c(1),c(2)))  \n',
             '    }} else {{ggOut =  \n',
             '      grid.arrange(ggOut, ggLeg, heights = c(7,2),  \n',
             '                   layout_matrix = rbind(c(1),c(2)))  \n',
             '    }}  \n',
             '  }} else {{ \n',
             '    if(inpRow & inpCol){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggCol, ggRow, widths = c(7,1), heights = c(1,7,2),  \n',
             '                  layout_matrix = rbind(c(3,NA),c(1,4),c(2,NA)))  \n',
             '    }} else if(inpRow){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggRow, widths = c(7,1), heights = c(7,2),  \n',
             '                  layout_matrix = rbind(c(1,3),c(2,NA)))  \n',
             '    }} else if(inpCol){{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, ggCol, heights = c(1,7,2),  \n',
             '                  layout_matrix = rbind(c(3),c(1),c(2)))  \n',
             '    }} else {{ggOut =  \n',
             '      arrangeGrob(ggOut, ggLeg, heights = c(7,2),  \n',
             '                  layout_matrix = rbind(c(1),c(2)))  \n',
             '    }}  \n',
             '  }} \n',
             '  return(ggOut) \n',
             '}} \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n',
             '### Start server code \n',
             'shinyServer(function(input, output, session) {{ \n',
             '  ### For all tags and Server-side selectize \n',
             '  observe_helpers() \n',
             ' \n')
}

#' Write code for main block of server.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrSVmain
#' @export wrSVmain
#'
wrSVmain <- function(prefix, subst = "") {
  glue::glue('optCrt="{{ option_create: function(data,escape) {{return(\'<div class=\\"create\\"><strong>\' + \'</strong></div>\');}} }}" \n', 
             '  updateSelectizeInput(session, "{prefix}a1inp2", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}a1cinp1", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}a3inp1", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}a3inp2", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene2, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b2inp1", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b2inp2", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene2, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b2inp3", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b2inp4", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene2, options = list( \n',
             '                         maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}c1inp2", server = TRUE, \n',
             '                       choices = c({prefix}conf[is.na(fID)]$UI,names({prefix}gene)), \n',
             '                       selected = {prefix}conf[is.na(fID)]$UI[1], options = list( \n',
             '                         maxOptions = length({prefix}conf[is.na(fID)]$UI) + 3, \n',
             '                         create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             ' \n',
             '  {prefix}b3spl_choices = list() \n',
             '  i=1 #; j=1 \n',
             '  for(index in strsplit({prefix}conf$fID, "\\\\|")) {{ \n',
             '    # if(is.na(index)[1]) {{ \n',
             '    #   i<-i+1 \n',
             '    #   next \n',
             '    # }} \n',
             '    # else {{ \n',
             '    #   if(length(index) == 2) {{ \n', 
             '    #     {prefix}b3spl_choices[[j]]<-{prefix}conf$ID[[i]] \n',
             '    #     j<-j+1 \n',
             '    #   }} \n',
             '    #   else {{ \n',
             '    #     {prefix}conf$split[[i]] <- {prefix}conf$ID[[i]] \n',
             '    #   }} \n',
             '    #   i<-i+1 \n',
             '    # }} \n',
             '    if(length(index) == 2) {{ \n',
             '      {prefix}b3spl_choices[[i]]<-{prefix}conf$ID[[i]] \n',
             '      i<-i+1 \n',
             '    }} \n',
             '  }} \n', 
             ' \n',
             '  updateSelectizeInput(session, "{prefix}b3spl", choices = {prefix}b3spl_choices, server = TRUE, \n',
             '                       selected = NULL, options = list( \n',
             '                       create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b3inp1", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene1, options = list( \n',
             '                       maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  updateSelectizeInput(session, "{prefix}b3inp2", choices = names({prefix}gene), server = TRUE, \n',
             '                       selected = {prefix}def$gene2, options = list( \n',
             '                       maxOptions = 7, create = TRUE, persist = TRUE, render = I(optCrt))) \n',
             '  \n',
             '  \n',
             '  ### Plots for tab a1 \n',
             '{subst}  output${prefix}a1sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}a1sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a1sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a1sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}a1oup1 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,  \n',
             '             input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '             input${prefix}a1siz, input${prefix}a1col1, input${prefix}a1ord1, \n',
             '             input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt, input${prefix}a1lab1, input${prefix}a1legend) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a1oup1", height = pList[input${prefix}a1psz]) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a1oup1.h, width = input${prefix}a1oup1.w, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,   \n',
             '                      input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '                      input${prefix}a1siz, input${prefix}a1col1, input${prefix}a1ord1,  \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt, input${prefix}a1lab1, input${prefix}a1legend) ) \n',
             '  }}) \n',
             '  output${prefix}a1oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a1oup1.h, width = input${prefix}a1oup1.w, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp1,   \n',
             '                      input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '                      input${prefix}a1siz, input${prefix}a1col1, input${prefix}a1ord1,  \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt, input${prefix}a1lab1, input${prefix}a1legend) ) \n',
             '  }}) \n',
             '  output${prefix}a1.dt <- renderDataTable({{ \n',
             '    ggData = scDRnum({prefix}conf, {prefix}meta, input${prefix}a1inp1, input${prefix}a1inp2, \n',
             '                     input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '                     "{prefix}gexpr.h5", {prefix}gene, input${prefix}a1splt) \n',
             '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '      formatRound(columns = c("pctExpress"), digits = 2) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a1oup2 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '             input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a1siz, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '             input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a1oup2", height = pList[input${prefix}a1psz]) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a1oup2.h, width = input${prefix}a1oup2.w, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '                      input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a1siz, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '  output${prefix}a1oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a1drX,"_",input${prefix}a1drY,"_",  \n',
             '                                   input${prefix}a1inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a1oup2.h, width = input${prefix}a1oup2.w, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1drX, input${prefix}a1drY, input${prefix}a1inp2,  \n',
             '                      input${prefix}a1sub1, input${prefix}a1sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a1siz, input${prefix}a1col2, input${prefix}a1ord2, \n',
             '                      input${prefix}a1fsz, input${prefix}a1asp, input${prefix}a1txt) ) \n',
             '  }}) \n',
             '   \n',
             '### Plots for tab a1b SPLIT CELLS \n',
             '{subst}  output${prefix}a1bsub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1bsub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}a1bsub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a1bsub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1sbub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a1bsub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a1bsub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a1bsub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a1bsub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             'output${prefix}a1boup1 <- renderPlot({{ \n',
             ' scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '          input${prefix}a1bsub1, input${prefix}a1bsub2, \n',
             '          input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '          input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=1, select=input${prefix}a1bsplit_select1_out2) \n',
             '}}) \n',
             'output${prefix}a1boup1.ui <- renderUI({{ \n',
             ' plotOutput("{prefix}a1boup1", height = pList[input${prefix}a1bpsz]) \n',
             '}}) \n',
             'output${prefix}a1boup1.pdf <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1bsplit1,"_",input${prefix}a1bsplit_select1_out2,"_", \n',
             '                                 input${prefix}a1binp1,".pdf") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "pdf", height = input${prefix}a1boup1.h, width = input${prefix}a1boup1.w, useDingbats = FALSE, \n',
             '   plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '   input${prefix}a1bsub1, input${prefix}a1bsub2, \n',
             '   input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '   input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=1, select=input${prefix}a1bsplit_select1_out2) ) \n',
             ' }}) \n',
             'output${prefix}a1boup1.png <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1bsplit1,"_",input${prefix}a1bsplit_select1_out2,"_", \n',
             '                                 input${prefix}a1binp1,".png") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "png", height = input${prefix}a1boup1.h, width = input${prefix}a1boup1.w, \n',
             '   plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '   input${prefix}a1bbsub1, input${prefix}a1bsub2, \n',
             '   input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '   input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=1, select=input${prefix}a1bsplit_select1_out2) ) \n',
             ' }}) \n',
             ' \n',
             'output${prefix}a1boup2 <- renderPlot({{ \n',
             ' scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '          input${prefix}a1bsub1, input${prefix}a1bsub2, \n',
             '          input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '          input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=2, select=input${prefix}a1bsplit_select2_out2) \n',
             '}}) \n',
             'output${prefix}a1boup2.ui <- renderUI({{ \n',
             ' plotOutput("{prefix}a1boup2", height = pList[input${prefix}a1bpsz]) \n',
             '}}) \n',
             'output${prefix}a1boup2.pdf <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1bsplit1,"_",input${prefix}a1bsplit_select2_out2,"_", \n',
             '                                 input${prefix}a1binp2,".pdf") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "pdf", height = input${prefix}a1boup2.h, width = input${prefix}a1boup2.w, useDingbats = FALSE, \n',
             '   plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '     input${prefix}a1bsub1, input${prefix}a1bsub2, \n',
             '     input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '     input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=2, select=input${prefix}a1bsplit_select2_out2)) \n',
             '}}) \n',
             'output${prefix}a1boup2.png <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1bsplit1,"_",input${prefix}a1bsplit_select2_out2,"_", \n',
             '                                 input${prefix}a1binp2,".png") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "png", height = input${prefix}a1boup2.h, width = input${prefix}a1boup2.w, \n',
             '   plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a1bdrX, input${prefix}a1bdrY, input${prefix}a1binp1, \n',
             '     input${prefix}a1bsub1, input${prefix}a1bsub2, \n',
             '     input${prefix}a1bsiz, input${prefix}a1bcol1, input${prefix}a1bord1, \n',
             '     input${prefix}a1bfsz, input${prefix}a1basp, input${prefix}a1btxt, input${prefix}a1blab1, input${prefix}a1blegend, insplit=input${prefix}a1bsplit1, split_idx=2, select=input${prefix}a1bsplit_select2_out2)) \n',
             '}}) \n',
             'output${prefix}a1bsplit_select1_out1 <- renderUI({{ \n',
             '  sub = strsplit({prefix}conf[UI == input${prefix}a1bsplit1]$fID, "\\\\|")[[1]] \n',
             '  selectInput("{prefix}a1bsplit_select1_out2", "Select state:", choices = sub, selected = sub[1]) \n',
             '}}) \n',
             'output${prefix}a1bsplit_select2_out1 <- renderUI({{ \n',
             '  sub = strsplit({prefix}conf[UI == input${prefix}a1bsplit1]$fID, "\\\\|")[[1]] \n',
             '  selectInput("{prefix}a1bsplit_select2_out2", "Select state:", choices = sub, selected = sub[2]) \n',
             ' }}) \n',
             ' \n',
             ' \n',
             ' \n',
             '### Plots for tab a1c SPLIT GENE EXP \n',
             'output${prefix}a1csub1.ui <- renderUI({{ \n',
             ' sub = strsplit({prefix}conf[UI == input${prefix}a1csub1]$fID, "\\\\|")[[1]] \n',
             ' checkboxGroupInput("{prefix}a1csub2", "Select which cells to show", inline = TRUE, \n',
             '                    choices = sub, selected = sub) \n',
             '}}) \n',
             'observeEvent(input${prefix}a1csub1non, {{ \n',
             ' sub = strsplit({prefix}conf[UI == input${prefix}a1csub1]$fID, "\\\\|")[[1]] \n',
             ' updateCheckboxGroupInput(session, inputId = "{prefix}a1csub2", label = "Select which cells to show", \n',
             '                          choices = sub, selected = NULL, inline = TRUE) \n',
             '}}) \n',
             'observeEvent(input${prefix}a1csub1all, {{ \n',
             ' sub = strsplit({prefix}conf[UI == input${prefix}a1csub1]$fID, "\\\\|")[[1]] \n',
             ' updateCheckboxGroupInput(session, inputId = "{prefix}a1csub2", label = "Select which cells to show", \n',
             '                          choices = sub, selected = sub, inline = TRUE) \n',
             '}}) \n',
             ' \n',
             'output${prefix}a1coup1 <- renderPlot({{ \n',
             ' scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=1, select=input${prefix}a1csplit_select1_out2) \n',
             '}}) \n',
             ' \n',
             'output${prefix}a1coup1.ui <- renderUI({{ \n',
             ' plotOutput("{prefix}a1coup1", height = pList[input${prefix}a1cpsz]) \n',
             '}}) \n',
             ' \n',
             ' \n',
             'output${prefix}a1coup1.pdf <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1csplit1,"_",input${prefix}a1csplit_select1_out2,"_", \n',
             '                                 input${prefix}a1cinp1,".pdf") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "pdf", height = input${prefix}a1coup1.h, width = input${prefix}a1coup1.w, useDingbats = FALSE, \n',
             '   plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=1, select=input${prefix}a1csplit_select1_out2) ) \n',
             '}}) \n',
             'output${prefix}a1coup1.png <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1csplit1,"_",input${prefix}a1csplit_select1_out2,"_", \n',
             '                                 input${prefix}a1cinp1,".png") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "png", height = input${prefix}a1coup1.h, width = input${prefix}a1coup1.w, \n',
             '   plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=1, select=input${prefix}a1csplit_select1_out2) ) \n',
             '}}) \n',
             'output${prefix}a1c.dt <- renderDataTable({{ \n',
             ' ggData = scDRnum({prefix}conf, {prefix}meta, input${prefix}a1cinp1, input${prefix}a1cinp2, \n',
             '                  input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '                  "{prefix}gexpr.h5", {prefix}gene, input${prefix}a1csplt) \n',
             ' datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '           options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '           formatRound(columns = c("pctExpress"), digits = 2) \n',
             '}}) \n',
             ' \n',
             'output${prefix}a1coup2 <- renderPlot({{ \n',
             ' scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=2, select=input${prefix}a1csplit_select2_out2) \n',
             '}}) \n',
             'output${prefix}a1coup2.ui <- renderUI({{ \n',
             ' plotOutput("{prefix}a1coup2", height = pList[input${prefix}a1cpsz]) \n',
             '}}) \n',
             ' \n',
             'output${prefix}a1coup2.pdf <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1csplit1,"_",input${prefix}a1csplit_select2_out2,"_", \n',
             '                                 input${prefix}a1cinp1,".pdf") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "pdf", height = input${prefix}a1coup2.h, width = input${prefix}a1coup2.w, useDingbats = FALSE, \n',
             '   plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=2, select=input${prefix}a1csplit_select2_out2) ) \n',
             '}}) \n',
             'output${prefix}a1coup2.png <- downloadHandler( \n',
             ' filename = function() {{ paste0("{prefix}","_",input${prefix}a1csplit1,"_",input${prefix}a1csplit_select2_out2,"_", \n',
             '                                 input${prefix}a1cinp1,".png") }}, \n',
             ' content = function(file) {{ ggsave( \n',
             '   file, device = "png", height = input${prefix}a1coup2.h, width = input${prefix}a1coup2.w, \n',
             '   plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a1cdrX, input${prefix}a1cdrY, input${prefix}a1cinp1, \n',
             '          input${prefix}a1csub1, input${prefix}a1csub2, \n',
             '          "{prefix}gexpr.h5", {prefix}gene, \n',
             '          input${prefix}a1csiz, input${prefix}a1ccol1, input${prefix}a1cord1, \n',
             '          input${prefix}a1cfsz, input${prefix}a1casp, input${prefix}a1ctxt, insplit=input${prefix}a1csplit1, split_idx=2, select=input${prefix}a1csplit_select2_out2) ) \n',
             '}}) \n',
             'output${prefix}a1csplit_select1_out1 <- renderUI({{ \n',
             '  sub = strsplit({prefix}conf[UI == input${prefix}a1csplit1]$fID, "\\\\|")[[1]] \n',
             '  selectInput("{prefix}a1csplit_select1_out2", "Select state:", choices = sub, selected = sub[1]) \n',
             '}}) \n',
             'output${prefix}a1csplit_select2_out1 <- renderUI({{ \n',
             '  sub = strsplit({prefix}conf[UI == input${prefix}a1csplit1]$fID, "\\\\|")[[1]] \n',
             '  selectInput("{prefix}a1csplit_select2_out2", "Select state:", choices = sub, selected = sub[2]) \n',
             '}}) \n',
             '   \n',
             '  ### Plots for tab a2 \n',
             '{subst}  output${prefix}a2sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}a2sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a2sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a2sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}a2oup1 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,  \n',
             '             input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '             input${prefix}a2siz, input${prefix}a2col1, input${prefix}a2ord1, \n',
             '             input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab1, input${prefix}a2legend1) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a2oup1", height = pList[input${prefix}a2psz]) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a2oup1.h, width = input${prefix}a2oup1.w, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,   \n',
             '                      input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '                      input${prefix}a2siz, input${prefix}a2col1, input${prefix}a2ord1,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab1, input${prefix}a2legend1) ) \n',
             '  }}) \n',
             '  output${prefix}a2oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a2oup1.h, width = input${prefix}a2oup1.w, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp1,   \n',
             '                      input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '                      input${prefix}a2siz, input${prefix}a2col1, input${prefix}a2ord1,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab1, input${prefix}a2legend1) ) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a2oup2 <- renderPlot({{ \n',
             '    scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,  \n',
             '             input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '             input${prefix}a2siz, input${prefix}a2col2, input${prefix}a2ord2, \n',
             '             input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab2, input${prefix}a2legend2) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a2oup2", height = pList[input${prefix}a2psz]) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a2oup2.h, width = input${prefix}a2oup2.w, useDingbats = FALSE, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,   \n',
             '                      input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '                      input${prefix}a2siz, input${prefix}a2col2, input${prefix}a2ord2,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab2, input${prefix}a2legend2) ) \n',
             '  }}) \n',
             '  output${prefix}a2oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a2drX,"_",input${prefix}a2drY,"_",  \n',
             '                                   input${prefix}a2inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a2oup2.h, width = input${prefix}a2oup2.w, \n',
             '      plot = scDRcell({prefix}conf, {prefix}meta, input${prefix}a2drX, input${prefix}a2drY, input${prefix}a2inp2,   \n',
             '                      input${prefix}a2sub1, input${prefix}a2sub2, \n',
             '                      input${prefix}a2siz, input${prefix}a2col2, input${prefix}a2ord2,  \n',
             '                      input${prefix}a2fsz, input${prefix}a2asp, input${prefix}a2txt, input${prefix}a2lab2, input${prefix}a2legend2) ) \n',
             '  }}) \n',
             '   \n',
             '   \n',
             '  ### Plots for tab a3 \n',
             '{subst}  output${prefix}a3sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}a3sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a3sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a3sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}a3sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}a3sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}a3sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}a3oup1 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '             input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a3siz, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '             input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a3oup1", height = pList[input${prefix}a3psz]) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp1,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a3oup1.h, width = input${prefix}a3oup1.w, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '                      input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '  output${prefix}a3oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp1,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a3oup1.h, width = input${prefix}a3oup1.w, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp1,  \n',
             '                      input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz, input${prefix}a3col1, input${prefix}a3ord1, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '   \n',
             '  output${prefix}a3oup2 <- renderPlot({{ \n',
             '    scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '             input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}a3siz, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '             input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}a3oup2", height = pList[input${prefix}a3psz]) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}a3oup2.h, width = input${prefix}a3oup2.w, useDingbats = FALSE, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '                      input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '  output${prefix}a3oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}a3drX,"_",input${prefix}a3drY,"_",  \n',
             '                                   input${prefix}a3inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}a3oup2.h, width = input${prefix}a3oup2.w, \n',
             '      plot = scDRgene({prefix}conf, {prefix}meta, input${prefix}a3drX, input${prefix}a3drY, input${prefix}a3inp2,  \n',
             '                      input${prefix}a3sub1, input${prefix}a3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}a3siz, input${prefix}a3col2, input${prefix}a3ord2, \n',
             '                      input${prefix}a3fsz, input${prefix}a3asp, input${prefix}a3txt) ) \n',
             '  }}) \n',
             '     \n',
             '   \n',
             '  ### Plots for tab b2 \n',
             '{subst}  output${prefix}b2sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}b2sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}b2sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}b2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}b2sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}b2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}b2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}b2oup1 <- renderPlot({{ \n',
             '    scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY,   \n',
             '             input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '             input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) \n',
             '  }}) \n',
             '  output${prefix}b2oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}b2oup1", height = pList2[input${prefix}b2psz]) \n',
             '  }}) \n',
             '  output${prefix}b2oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_",  \n',
             '                                    input${prefix}b2inp1,"_",input${prefix}b2inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}b2oup1.h, width = input${prefix}b2oup1.w, useDingbats = FALSE, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY,  \n',
             '                      input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '                      input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) ) \n',
             '  }}) \n',
             '  output${prefix}b2oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_",  \n',
             '                                    input${prefix}b2inp1,"_",input${prefix}b2inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}b2oup1.h, width = input${prefix}b2oup1.w, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY,  \n',
             '                      input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '                      input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) ) \n',
             '  }}) \n',
             '  output${prefix}b2oup2 <- renderPlot({{ \n',
             '    scDRcoexLeg(input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2col1, input${prefix}b2fsz) \n',
             '  }}) \n',
             '  output${prefix}b2oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}b2oup2", height = "300px") \n',
             '  }}) \n',
             '  output${prefix}b2oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_",  \n',
             '                                    input${prefix}b2inp1,"_",input${prefix}b2inp2,"_leg.pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 3, width = 4, useDingbats = FALSE, \n',
             '      plot = scDRcoexLeg(input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2col1, input${prefix}b2fsz) ) \n',
             '  }}) \n',
             '  output${prefix}b2oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_",  \n',
             '                                    input${prefix}b2inp1,"_",input${prefix}b2inp2,"_leg.png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 3, width = 4, \n',
             '      plot = scDRcoexLeg(input${prefix}b2inp1, input${prefix}b2inp2, input${prefix}b2col1, input${prefix}b2fsz) ) \n',
             '  }}) \n',
             '  output${prefix}b2.dt <- renderDataTable({{ \n',
             '    ggData = scDRcoexNum({prefix}conf, {prefix}meta, input${prefix}b2inp1, input${prefix}b2inp2, \n',
             '                         input${prefix}b2sub1, input${prefix}b2sub2, "{prefix}gexpr.h5", {prefix}gene) \n',
             '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '      formatRound(columns = c("percent"), digits = 2) \n',
             '  }}) \n',
             '  output${prefix}b2oup3 <- renderPlot({{ \n',
             '    scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY, \n',
             '            input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '            "{prefix}gexpr.h5", {prefix}gene, \n',
             '            input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '            input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) \n',
             '  }}) \n',
             '  output${prefix}b2oup3.ui <- renderUI({{ \n', 
             '    plotOutput("{prefix}b2oup3", height = pList2[input${prefix}b2psz]) \n',
             '  }}) \n',
             '  output${prefix}b2oup3.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_", \n', 
             '                                    input${prefix}b2inp3,"_",input${prefix}b2inp4,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}b2oup3.h, width = input${prefix}b2oup3.w, useDingbats = FALSE, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY, \n',
             '                      input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '                      input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) ) \n',
             '  }}) \n',
             '  output${prefix}b2oup3.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_", \n', 
             '                                    input${prefix}b2inp3,"_",input${prefix}b2inp4,".png") }}, \n',
             '  content = function(file) {{ ggsave( \n',
             '    file, device = "png", height = input${prefix}b2oup3.h, width = input${prefix}b2oup3.w, \n',
             '    plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b2drX, input${prefix}b2drY, \n', 
             '                    input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2sub1, input${prefix}b2sub2, \n',
             '                    "{prefix}gexpr.h5", {prefix}gene, \n',
             '                    input${prefix}b2siz, input${prefix}b2col1, input${prefix}b2ord1, \n',
             '                    input${prefix}b2fsz, input${prefix}b2asp, input${prefix}b2txt) ) \n',
             '   }}) \n',
             '  output${prefix}b2oup4 <- renderPlot({{ \n',
             '    scDRcoexLeg(input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2col1, input${prefix}b2fsz) \n',
             '  }}) \n', 
             '  output${prefix}b2oup4.ui <- renderUI({{ \n', 
             '    plotOutput("{prefix}b2oup4", height = "300px") \n',
             '  }}) \n',
             '  output${prefix}b2oup4.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_", \n',
             '                                    input${prefix}b2inp3,"_",input${prefix}b2inp4,"_leg.pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 3, width = 4, useDingbats = FALSE, \n',
             '      plot = scDRcoexLeg(input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2col1, input${prefix}b2fsz) ) \n',
             '  }}) \n',
             '  output${prefix}b2oup4.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b2drX,"_",input${prefix}b2drY,"_", \n',
             '                                    input${prefix}b2inp3,"_",input${prefix}b2inp4,"_leg.png") }, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 3, width = 4, \n',
             '      plot = scDRcoexLeg(input${prefix}b2inp3, input${prefix}b2inp4, input${prefix}b2col1, input${prefix}b2fsz) ) \n',
             '  }}) \n',
             '  output${prefix}b2_2.dt <- renderDataTable({{ \n', 
             '    ggData = scDRcoexNum({prefix}conf, {prefix}meta, input${prefix}b2inp3, input${prefix}b2inp4, \n',
             '                          input${prefix}b2sub1, input${prefix}b2sub2, "{prefix}gexpr.h5", {prefix}gene) \n',
             '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '              options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '      formatRound(columns = c("percent"), digits = 2) \n',
             '  }}) \n',
             '   \n',
             '   \n',
             '   \n',
             '  ### b3 \n',
             '  output${prefix}b3sub1.ui <- renderUI({{ \n',
             '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
             '    checkboxGroupInput("{prefix}b3sub2", "Select which cells to show", inline = TRUE, \n',
             '                       choices = sub, selected = sub) \n',
             '  }}) \n',
             '  observeEvent(input${prefix}b3sub1non, {{ \n', 
             '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
             '    updateCheckboxGroupInput(session, inputId = "{prefix}b3sub2", label = "Select which cells to show", \n',
             '                             choices = sub, selected = NULL, inline = TRUE) \n',
             '  }}) \n', 
             '  observeEvent(input${prefix}b3sub1all, {{ \n', 
             '    sub = strsplit({prefix}conf[UI == input${prefix}b3sub1]$fID, "\\\\|")[[1]] \n',
             '    updateCheckboxGroupInput(session, inputId = "{prefix}b3sub2", label = "Select which cells to show", \n',
             '                             choices = sub, selected = sub, inline = TRUE) \n',
             '  }}) \n', 
             '  output${prefix}b3oup1 <- renderPlot({{ \n', 
             '    #split_param_1 <- strsplit({prefix}conf[{prefix}conf$split==TRUE]$fID, "\\\\|")[[1]][1] \n',
             '    scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '              input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '              "{prefix}gexpr.h5", {prefix}gene, \n',
             '              input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '              input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select1_out2) \n',
             '  }}) \n', 
             '  output${prefix}b3oup1.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}b3oup1", height = pList2[input${prefix}b3psz]) \n',
             '  }}) \n', 
             '  output${prefix}b3oup1.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}","_",input${prefix}b3inp1,"_",input${prefix}b3inp2,"_", \n',
             '                                    input${prefix}b3split,"_",input${prefix}b3split_select1_out2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}b3oup1.h, width = input${prefix}b3oup1.w, useDingbats = FALSE, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '                      input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '                      input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select1_out2) ) \n',
             '  }}) \n',
             '  output${prefix}b3oup1.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}","_",input${prefix}b3inp1,"_",input${prefix}b3inp2,"_", \n',
             '                                    input${prefix}b3split,"_",input${prefix}b3split_select1_out2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}b3oup1.h, width = input${prefix}b3oup1.w, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '                      input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '                      input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select1_out2) ) \n',
             '  }}) \n', 
             '  output${prefix}b3oup1.dt <- renderDataTable({{ \n',
             '    #split_param_1 <- strsplit({prefix}conf[{prefix}conf$split==TRUE]$fID, "\\\\|")[[1]][1] \n',
             '    ggData = scDRcoexNum({prefix}conf, {prefix}meta, input${prefix}b3inp1, input${prefix}b3inp2, \n',
             '                         input${prefix}b3sub1, input${prefix}b3sub2, "{prefix}gexpr.h5", {prefix}gene, input${prefix}b3split, input${prefix}b3split_select1_out2) \n',
             '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '      options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '      formatRound(columns = c("percent"), digits = 2) \n',
             '  }}) \n', 
             '  output${prefix}b3oup2 <- renderPlot({{ \n',
             '    #split_param_2 <- strsplit({prefix}conf[{prefix}conf$split==TRUE]$fID, "\\\\|")[[1]][2] \n',
             '    scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '             input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, \n',
             '             input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '             input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select2_out2) \n',
             '  }}) \n',
             '  output${prefix}b3oup2.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}b3oup2", height = pList2[input${prefix}b3psz]) \n',
             '  }}) \n', 
             '  output${prefix}b3oup2.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}","_",input${prefix}b3inp1,"_",input${prefix}b3inp2,"_", \n',
             '                                    input${prefix}b3split,"_",input${prefix}b3split_select2_out2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}b3oup2.h, width = input${prefix}b3oup2.w, useDingbats = FALSE, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '                      input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '                      input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select2_out2) ) \n',
             '  }}) \n', 
             '  output${prefix}b3oup2.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}","_",input${prefix}b3inp1,"_",input${prefix}b3inp2,"_", \n',
             '                                    input${prefix}b3split,"_",input${prefix}b3split_select2_out2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}b3oup2.h, width = input${prefix}b3oup2.w, \n',
             '      plot = scDRcoex({prefix}conf, {prefix}meta, input${prefix}b3drX, input${prefix}b3drY, \n',
             '                      input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3sub1, input${prefix}b3sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, \n',
             '                      input${prefix}b3siz, input${prefix}b3col1, input${prefix}b3ord1, \n',
             '                      input${prefix}b3fsz, input${prefix}b3asp, input${prefix}b3txt, input${prefix}b3split, input${prefix}b3split_select2_out2) ) \n',
             '  }}) \n',
             '  output${prefix}b3oup2.dt <- renderDataTable({{ \n', 
             '    #split_param_2 <- strsplit({prefix}conf[{prefix}conf$split==TRUE]$fID, "\\\\|")[[1]][2] \n',
             '    ggData = scDRcoexNum({prefix}conf, {prefix}meta, input${prefix}b3inp1, input${prefix}b3inp2, \n',
             '                         input${prefix}b3sub1, input${prefix}b3sub2, "{prefix}gexpr.h5", {prefix}gene, input${prefix}b3split, input${prefix}b3split_select2_out2) \n',
             '    datatable(ggData, rownames = FALSE, extensions = "Buttons", \n',
             '      options = list(pageLength = -1, dom = "tB", buttons = c("copy", "csv", "excel"))) %>% \n',
             '      formatRound(columns = c("percent"), digits = 2) \n',
             '  }}) \n', 
             '  output${prefix}b3oup3 <- renderPlot({{ \n', 
             '    scDRcoexLeg(input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3col1, input${prefix}b3fsz) \n',
             '  }}) \n', 
             '  output${prefix}b3oup3.ui <- renderUI({{ \n', 
             '    plotOutput("{prefix}b3oup3", height = "300px") \n',
             '  }}) \n', 
             '  output${prefix}b3oup3.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b3drX,"_",input${prefix}b3drY,"_", \n',
             '                                    input${prefix}b3inp1,"_",input${prefix}b3inp2,"_leg.pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = 3, width = 4, useDingbats = FALSE, \n',
             '      plot = scDRcoexLeg(input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3col1, input${prefix}b3fsz) ) \n',
             '  }}) \n', 
             '  output${prefix}b3oup3.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}b3drX,"_",input${prefix}b3drY,"_", \n',
             '                                    input${prefix}b3inp1,"_",input${prefix}b3inp2,"_leg.png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = 3, width = 4, \n',
             '      plot = scDRcoexLeg(input${prefix}b3inp1, input${prefix}b3inp2, input${prefix}b3col1, input${prefix}b3fsz) ) \n',
             '  }}) \n',
             '  output${prefix}b3split_select1_out1 <- renderUI({{ \n', 
             '    sub = strsplit({prefix}conf[UI == input${prefix}b3split]$fID, "\\\\|")[[1]] \n',
             '    selectInput("{prefix}b3split_select1_out2", "Select state:", choices = sub, selected = sub[1]) \n',
             '  }}) \n', 
             '  output${prefix}b3split_select2_out1 <- renderUI({{ \n',
             '    sub = strsplit({prefix}conf[UI == input${prefix}b3split]$fID, "\\\\|")[[1]] \n',
             '    selectInput("{prefix}b3split_select2_out2", "Select state:", choices = sub, selected = sub[2]) \n',
             '  }}) \n', 
             '  \n',
             '  ### Plots for tab c1 \n',
             '{subst}  output${prefix}c1sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}c1sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}c1sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}c1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}c1sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}c1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}c1oup <- renderPlot({{ \n',
             '    scVioBox({prefix}conf, {prefix}meta, input${prefix}c1inp1, input${prefix}c1inp2, \n',
             '             input${prefix}c1sub1, input${prefix}c1sub2, \n',
             '             "{prefix}gexpr.h5", {prefix}gene, input${prefix}c1typ, input${prefix}c1pts, \n',
             '             input${prefix}c1siz, input${prefix}c1fsz) \n',
             '  }}) \n',
             '  output${prefix}c1oup.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}c1oup", height = pList2[input${prefix}c1psz]) \n',
             '  }}) \n',
             '  output${prefix}c1oup.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}c1typ,"_",input${prefix}c1inp1,"_",  \n',
             '                                   input${prefix}c1inp2,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}c1oup.h, width = input${prefix}c1oup.w, useDingbats = FALSE, \n',
             '      plot = scVioBox({prefix}conf, {prefix}meta, input${prefix}c1inp1, input${prefix}c1inp2, \n',
             '                      input${prefix}c1sub1, input${prefix}c1sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, input${prefix}c1typ, input${prefix}c1pts, \n',
             '                      input${prefix}c1siz, input${prefix}c1fsz) ) \n',
             '  }}) \n',
             '  output${prefix}c1oup.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}c1typ,"_",input${prefix}c1inp1,"_",  \n',
             '                                   input${prefix}c1inp2,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}c1oup.h, width = input${prefix}c1oup.w, \n',
             '      plot = scVioBox({prefix}conf, {prefix}meta, input${prefix}c1inp1, input${prefix}c1inp2, \n',
             '                      input${prefix}c1sub1, input${prefix}c1sub2, \n',
             '                      "{prefix}gexpr.h5", {prefix}gene, input${prefix}c1typ, input${prefix}c1pts, \n',
             '                      input${prefix}c1siz, input${prefix}c1fsz) ) \n',
             '  }}) \n',
             '     \n',
             '   \n', 
             '### Plots for tab c2 \n',
             '{subst}  output${prefix}c2sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}c2sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}c2sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}c2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}c2sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}c2sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}c2sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             'output${prefix}c2oup <- renderPlot({{ \n',
             '  scProp({prefix}conf, {prefix}meta, input${prefix}c2inp1, input${prefix}c2inp2,  \n',
             '         input${prefix}c2sub1, input${prefix}c2sub2, \n',
             '         input${prefix}c2typ, input${prefix}c2flp, input${prefix}c2fsz) \n',
             '}}) \n',
             'output${prefix}c2oup.ui <- renderUI({{ \n',
             '  plotOutput("{prefix}c2oup", height = pList2[input${prefix}c2psz]) \n',
             '}}) \n',
             'output${prefix}c2oup.pdf <- downloadHandler( \n',
             '  filename = function() {{ paste0("{prefix}",input${prefix}c2typ,"_",input${prefix}c2inp1,"_",  \n',
             '                                 input${prefix}c2inp2,".pdf") }}, \n',
             '  content = function(file) {{ ggsave( \n',
             '    file, device = "pdf", height = input${prefix}c2oup.h, width = input${prefix}c2oup.w, useDingbats = FALSE, \n',
             '    plot = scProp({prefix}conf, {prefix}meta, input${prefix}c2inp1, input${prefix}c2inp2,  \n',
             '                  input${prefix}c2sub1, input${prefix}c2sub2, \n',
             '                  input${prefix}c2typ, input${prefix}c2flp, input${prefix}c2fsz) ) \n',
             '  }}) \n',
             'output${prefix}c2oup.png <- downloadHandler( \n',
             '  filename = function() {{ paste0("{prefix}",input${prefix}c2typ,"_",input${prefix}c2inp1,"_",  \n',
             '                                 input${prefix}c2inp2,".png") }}, \n',
             '  content = function(file) {{ ggsave( \n',
             '    file, device = "png", height = input${prefix}c2oup.h, width = input${prefix}c2oup.w, \n',
             '    plot = scProp({prefix}conf, {prefix}meta, input${prefix}c2inp1, input${prefix}c2inp2,  \n',
             '                  input${prefix}c2sub1, input${prefix}c2sub2, \n',
             '                  input${prefix}c2typ, input${prefix}c2flp, input${prefix}c2fsz) ) \n',
             '  }}) \n',
             '     \n',
             '   \n', 
             '  ### Plots for tab d1 \n',
             '{subst}  output${prefix}d1sub1.ui <- renderUI({{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}d1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    checkboxGroupInput("{prefix}d1sub2", "Select which cells to show", inline = TRUE, \n',
             '{subst}                       choices = sub, selected = sub) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}d1sub1non, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}d1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}d1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = NULL, inline = TRUE) \n',
             '{subst}  }}) \n',
             '{subst}  observeEvent(input${prefix}d1sub1all, {{ \n',
             '{subst}    sub = strsplit({prefix}conf[UI == input${prefix}d1sub1]$fID, "\\\\|")[[1]] \n',
             '{subst}    updateCheckboxGroupInput(session, inputId = "{prefix}d1sub2", label = "Select which cells to show", \n',
             '{subst}                             choices = sub, selected = sub, inline = TRUE) \n',
             '{subst}  }}) \n',
             '  output${prefix}d1oupTxt <- renderUI({{ \n',
             '    geneList = scGeneList(input${prefix}d1inp, {prefix}gene) \n',
             '    if(nrow(geneList) > 50){{ \n',
             '      HTML("More than 50 input genes! Please reduce the gene list!") \n',
             '    }} else {{ \n',
             '      oup = paste0(nrow(geneList[present == TRUE]), " genes OK and will be plotted") \n',
             '      if(nrow(geneList[present == FALSE]) > 0){{ \n',
             '        oup = paste0(oup, "<br/>", \n',
             '                     nrow(geneList[present == FALSE]), " genes not found (", \n',
             '                     paste0(geneList[present == FALSE]$gene, collapse = ", "), ")") \n',
             '      }} \n',
             '      HTML(oup) \n',
             '    }} \n',
             '  }}) \n',
             '  output${prefix}d1oup <- renderPlot({{ \n',
             '    scBubbHeat({prefix}conf, {prefix}meta, input${prefix}d1inp, input${prefix}d1grp, input${prefix}d1plt, \n',
             '               input${prefix}d1sub1, input${prefix}d1sub2, "{prefix}gexpr.h5", {prefix}gene, \n',
             '               input${prefix}d1scl, input${prefix}d1row, input${prefix}d1col, \n',
             '               input${prefix}d1cols, input${prefix}d1fsz) \n',
             '  }}) \n',
             '  output${prefix}d1oup.ui <- renderUI({{ \n',
             '    plotOutput("{prefix}d1oup", height = pList3[input${prefix}d1psz]) \n',
             '  }}) \n',
             '  output${prefix}d1oup.pdf <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}d1plt,"_",input${prefix}d1grp,".pdf") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "pdf", height = input${prefix}d1oup.h, width = input${prefix}d1oup.w, \n',
             '      plot = scBubbHeat({prefix}conf, {prefix}meta, input${prefix}d1inp, input${prefix}d1grp, input${prefix}d1plt, \n',
             '                        input${prefix}d1sub1, input${prefix}d1sub2, "{prefix}gexpr.h5", {prefix}gene, \n',
             '                        input${prefix}d1scl, input${prefix}d1row, input${prefix}d1col, \n',
             '                        input${prefix}d1cols, input${prefix}d1fsz, save = TRUE) ) \n',
             '  }}) \n',
             '  output${prefix}d1oup.png <- downloadHandler( \n',
             '    filename = function() {{ paste0("{prefix}",input${prefix}d1plt,"_",input${prefix}d1grp,".png") }}, \n',
             '    content = function(file) {{ ggsave( \n',
             '      file, device = "png", height = input${prefix}d1oup.h, width = input${prefix}d1oup.w, \n',
             '      plot = scBubbHeat({prefix}conf, {prefix}meta, input${prefix}d1inp, input${prefix}d1grp, input${prefix}d1plt, \n',
             '                        input${prefix}d1sub1, input${prefix}d1sub2, "{prefix}gexpr.h5", {prefix}gene, \n',
             '                        input${prefix}d1scl, input${prefix}d1row, input${prefix}d1col, \n',
             '                        input${prefix}d1cols, input${prefix}d1fsz, save = TRUE) ) \n',
             '  }}) \n',
             '  \n',
             '  \n',
             '  \n')
}

#' Insert data server cues for seurat cluster markers, all 
#'
#' @param prefix file prefix
#' @param markers.all TRUE/FALSE whether to create this tab
#' @rdname wrSVmarkersAll
#' @export wrSVmarkersAll
#'
wrSVmarkersAll <- function(prefix, markers.all) {
  #graphs <- ''
  if(markers.all == TRUE) {
    #graphs <- graphs + glue::glue(
    glue::glue(
            '   # markers.all \n',
            '   output${prefix}m_all.ui <- renderUI ({{ \n',
            '     if({prefix}conf$markers[1] == "ERROR") {{ \n',
            '       renderText("ERROR: a new \'{prefix}m_all.rds\' was not created for this ShinyCellPlus instance!") \n',
            '     }} \n',
            '     else if({prefix}m_all[1,1] == "ERROR") {{ \n',
            '       renderText("ERROR: data file \'{prefix}m_all.rds\' not found, cannot create plots!") \n',
            '     }} \n',
            '     else {{ \n',
            '       DT::renderDataTable(subset({prefix}m_all, cluster == input${prefix}m_all_select), filter="top") \n',
            '     }} \n',
            '   }}) \n',
            ' \n',
            '   output${prefix}m_all_select.ui <- renderUI ({{ \n',
            '     if({prefix}conf$markers[1] == "ERROR" | {prefix}m_all[1,1] == "ERROR") {{ \n',
            '       renderText("") \n',
            '     }} \n',
            '     else {{ \n',
            '       renderUI({{ \n',
            '         selectInput("{prefix}m_all_select", "Select cell:", choices=unique({prefix}m_all$cluster)) \n',
            '       }}) \n',
            '     }} \n',
            '   }}) \n',
            ' \n'
    )
  }
  else {
    return('')
  }
}

#' Insert data server cues for seurat cluster markers, top 20 
#'
#' @param prefix file prefix
#' @param markers.top20 TRUE/FALSE whether to create this tab
#' @rdname wrSVmarkersTop20
#' @export wrSVmarkersTop20
#'
wrSVmarkersTop20 <- function(prefix, markers.top20) {
  graphs<-''
  if(markers.top20 == TRUE) {
    graphs <- graphs + glue::glue(
             '  # markers.top20 \n',
             '  output${prefix}m_t20.ui <- renderUI({{ \n',
             '    if({prefix}conf$markers[2] == "ERROR") {{ \n',
             '      renderText("ERROR: a new \'{prefix}m_t20.rds\' was not created for this ShinyCellPlus instance!") \n',
             '    }} \n',
             '    else if({prefix}m_t20[1,1] == "ERROR") {{ \n',
             '      renderText("ERROR: data file \'{prefix}m_t20.rds\' not found, cannot create plots!") \n',
             '    }} \n',
             '    else {{ \n',
             '      DT::renderDataTable({prefix}m_t20, filter="top") \n',
             '    }} \n',
             '  }}) \n',
             '  \n', 
             '  \n'
    )
  }
  return(graphs)
}

#' Insert data server cues for differential expression genes table tab
#'
#' @param prefix file prefix
#' @param de.genes TRUE/FALSE whether to create this tab
#' @rdname wrSVdeGenes
#' @export wrSVdeGenes
#'
wrSVdeGenes <- function(prefix, de.genes) {
  #graphs<-''
  if(de.genes == TRUE) {
    #graphs <- graphs + glue::glue(
    glue::glue(
            '   # de.genes \n',
            '   output${prefix}de_genes.ui <- renderUI ({{ \n',
            '     if({prefix}conf$DEs[1] == "ERROR") {{ \n',
            '       renderText("ERROR: a new \'{prefix}de_genes.rds\' was not created for this ShinyCellPlus instance!") \n',
            '     }} \n',
            '     else if({prefix}de_genes[1,1] == "ERROR") {{ \n',
            '       renderText("ERROR: data file \'{prefix}de_genes.rds\' not found, cannot create plots!") \n',
            '     }} \n',
            '     else {{ \n',
            '       #{prefix}de_genes <- readRDS("./{prefix}de_genes.rds") \n',
            '       #print({prefix}de_genes) \n',
            '       renderDataTable(subset({prefix}de_genes, de_name == input${prefix}de_genes_select), filter="top") \n',
            '     }} \n',
            '   }}) \n',
            '   \n',
            ' \n'
    )
  }
  else {
    return('')
  }
  #return(graphs)
}

#' Insert data server cues for gene signature manipulation via AUCell tab
#'
#' @param prefix file prefix
#' @param gene.ranks TRUE/FALSE whether to create this tab
#' @rdname wrSVgeneSig
#' @export wrSVgeneSig
#'
wrSVgeneSig <- function(prefix, gene.ranks) {
  #graphs <- ''
  if(gene.ranks == TRUE) {
    #  graphs<-graphs + glue::glue( 
    glue::glue(
            ' \n',
            ' # gene sig \n',
            ' output$gsig_subset.ui <- renderUI({{ \n',
            '   subset = strsplit({prefix}conf[UI == input$gsig_subset]$fID, "\\\\|")[[1]] \n',
            '   checkboxGroupInput("gsig_subset2", "Select which cells to show", inline = TRUE, \n',
            '                       choices = subset, selected = subset) \n',
            ' }}) \n',
            ' \n',
            ' observeEvent(input$gsig_subset_sel_all, {{ \n',
            '   subset = strsplit({prefix}conf[UI == input$gsig_subset]$fID, "\\\\|")[[1]] \n',
            '   updateCheckboxGroupInput(session, inputId = "gsig_subset2", label = "Select which cells to show", \n',
            '                             choices = subset, selected = subset, inline = TRUE) \n',
            ' }}) \n',
            ' \n',
            ' observeEvent(input$gsig_subset_desel_all, {{ \n',
            '   subset = strsplit({prefix}conf[UI == input$gsig_subset]$fID, "\\\\|")[[1]] \n',
            '   updateCheckboxGroupInput(session, inputId = "gsig_subset2", label = "Select which cells to show", \n',
            '                             choices = subset, selected = NULL, inline = TRUE) \n',
            ' }}) \n',
            ' \n',
            ' #TODO: redo utilizing R\'s vectors to save lines and iterations \n',
            ' output$gsig_list.html <- renderUI({{ \n',
            '   n_g_found<-0 \n',
            '   n_g_not_found<-0 \n',
            '   gs_not_found<-"" \n',
            '   output1<-"" \n',
            '   output2<-"" \n',
            '   if(input$gsig_list == "") {{ \n',
            '     output1 <- "0 genes found in the dataset" \n',
            '     return(output1) \n',
            '   }} \n',
            '   \n',
            '   for(i in strsplit(input$gsig_list, "\\n")[[1]]) {{ \n',
            '     if(!is.na({prefix}gene[i])) {{ \n',
            '       n_g_found<-n_g_found+1 \n',
            '     }} \n',
            '     else {{ \n',
            '       n_g_not_found<-n_g_not_found+1 \n',
            '       if(gs_not_found == "") {{ \n',
            '         gs_not_found<-sprintf("%s", i) \n',
            '       }} \n',
            '       else {{ \n',
            '         gs_not_found<-sprintf("%s, %s", gs_not_found, i) \n',
            '       }} \n',
            '     }} \n',
            '   }} \n',
            '   output1<-sprintf("%s genes found in the dataset", n_g_found) \n',
            '   if(n_g_not_found > 0) {{ \n',
            '     output2<-sprintf(", %s genes not found (%s)", n_g_not_found, gs_not_found) \n',
            '   }} \n',
            '   \n',
            '   renderText(c(output1, output2), sep="\\n") \n',
            ' }}) \n',
            ' \n',
            ' scGeneSig <- function(genes_list, sort_grouping, grouping_subset, grouping_subset_selections, subset_toggle, \n',
            '                       point_size, font_size, graphics_color, cell_info) {{ \n',
            '   genesets<-GSEABase::GeneSet( as.vector(strsplit(genes_list, "\\n"))[[1]], setName="selected_genes") \n',
            '   cells_auc<-AUCell_calcAUC(genesets, {prefix}gene_ranks, aucMaxRank=nrow({prefix}gene_ranks)*0.05) \n',
            '   cells_assignment<-AUCell_exploreThresholds(cells_auc, plotHist=TRUE, nCores=1, assignCells=TRUE) #more cores? \n',
            '   selected_genes<-getAUC(cells_auc) \n',
            '   selected_genes<-t(selected_genes) \n',
            '   meta<-cbind({prefix}meta, selected_genes) \n',
            '   \n',
            '   #take data from meta and assign to default columns for stable reference \n',
            '   meta[["selected_group"]] <- meta[[sort_grouping]] \n',
            '   meta[["subset"]] <- meta[[grouping_subset]] \n',
            '   if(subset_toggle != 0) {{ # shotty hotfix for blank generated graph when subset menu hasn\'t been opened yet... \n',
            '     meta <- meta[subset %in% grouping_subset_selections] \n',
            '   }} \n',
            '   \n',
            '   violins<-ggplot(meta, aes(x=selected_group, y=selected_genes, fill=selected_group)) + geom_violin(scale="width") + \n',
            '   geom_jitter(size=point_size, shape = 16, color="black", alpha=0.25) + \n',
            '   sctheme(base_size=sList2[[font_size]], Xang = 45, XjusH = 1) + \n',
            '   xlab(sort_grouping) + \n',
            '   ylab("Gene set AUC (per cell)") + \n',
            '   theme(legend.position = "none") \n',
            '   \n',
            '   if(any(names(meta) == "umap_1")) {{ \n',
            '     umap<-ggplot(meta, aes(x=umap_1, y=umap_2, color=selected_genes)) + geom_point(size=0.5) + \n',
            '       scale_color_gradientn("", colours = cList[[graphics_color]]) + \n',
            '       xlab("umap_1") + \n',
            '       ylab("umap_2") \n',
            '     \n',
            '     #UMAP labels toggle \n',
            '     if(cell_info) {{ \n',
            '       ggData3 <- meta[, .(X = mean(umap_1), Y = mean(umap_2)), by = sort_grouping] \n',
            '       lListX <- min(nchar(paste0(ggData3[[sort_grouping]], collapse = "")), 200) \n',
            '       lListX <- lList - (0.25 * floor(lListX/50)) \n',
            '       umap <- umap + \n',
            '       geom_text_repel(data = ggData3, aes(X, Y, label = ggData3[[sort_grouping]]), \n',
            '                       color = "grey10", bg.color = "grey95", bg.r = 0.15, \n',
            '                       size = lListX[font_size], seed = 42) \n',
            '     }} \n',
            '   }} \n',
            '   # for older seurat objects \n',
            '   else {{ # if(any(names(meta) == "UMAP_1") \n',
            '     umap<-ggplot(meta, aes(x=UMAP_1, y=UMAP_2, color=selected_genes)) + geom_point(size=0.5) + \n',
            '       scale_color_gradientn("", colours = cList[[graphics_color]]) + \n',
            '       xlab("UMAP_1") + \n',
            '       ylab("UMAP_2") \n',
            '     \n',
            '     if(cell_info) {{ \n', 
            '       ggData3 <- meta[, .(X = mean(UMAP_1), Y = mean(UMAP_2)), by = sort_grouping] \n',
            '       lListX <- min(nchar(paste0(ggData3[[sort_grouping]], collapse = "")), 200) \n',
            '       lListX <- lList - (0.25 * floor(lListX/50)) \n',
            '       umap <- umap + \n',
            '         geom_text_repel(data = ggData3, aes(X, Y, label = ggData3[[sort_grouping]]), \n',
            '                         color = "grey10", bg.color = "grey95", bg.r = 0.15, \n',
            '                         size = lListX[font_size], seed = 42) \n',
            '     }} \n',
            '     \n',
            '     #else: throw error plot \n',
            '   }} \n',
            '   \n',
            '   umap <- umap + theme_classic() \n',
            '   \n',
            '   ggOut <- grid.arrange(violins, umap, ncol=2) \n', 
            '   \n',
            '   return(ggOut)\n',
            ' }} \n',
            ' \n',
            ' output$gsig_plot <- renderPlot({{ \n',
            '   # if seurat@misc$gene_ranks$aucell$all was NULL upon site creation \n',
            '   if({prefix}conf$gene_ranks[1] == "ERROR") {{ \n',
            '     ggplot() + labs(title="ERROR: a new \'{prefix}gene_ranks.rds\' was not created for this ShinyCellPlus instance!") \n',
            '   }} \n',
            '   # if {prefix}gene_ranks.rds not found \n',
            '   else if(class({prefix}gene_ranks) == "matrix") {{ # flimsy check, need to find better way \n',
            '     if({prefix}gene_ranks[1,1] == "ERROR") {{ \n',
            '       ggplot() + labs(title="ERROR: data file \'{prefix}gene_ranks.rds\' not found, cannot create plots!") \n',
            '     }} \n',
            '   }} \n',
            '   else {{ \n',
            '     scGeneSig(input$gsig_list, input$gsig_group, input$gsig_subset, input$gsig_subset2, \n',
            '               input$gsig_subset_toggle, input$gsig_graphics_point_size, input$gsig_graphics_font_size, \n',
            '               input$gsig_graphics_color, input$gsig_graphics_cell_info) \n',
            '   }} \n',
            ' }}) \n',
            ' \n',
            ' output$gsig_download.pdf <- downloadHandler( \n',
            '   filename = function() {{ \n',
            '     paste0("{prefix}_gene_sig_", Sys.Date(), ".pdf") \n',
            '   }}, \n',
            '   content = function(file) {{ ggsave( \n',
            '     file, device = "pdf", height = input$gsig_download_height, width = input$gsig_download_width, useDingbats = FALSE,\n',
            '     plot = scGeneSig(input$gsig_list, input$gsig_group, input$gsig_subset, input$gsig_subset2, \n',
            '         input$gsig_subset_toggle, input$gsig_graphics_point_size, input$gsig_graphics_font_size, \n',
            '         input$gsig_graphics_color, input$gsig_graphics_cell_info) \n',
            '     ) \n',
            '   }} \n',
            ' ) \n',
            ' \n',
            ' output$gsig_download.png <- downloadHandler( \n',
            '   filename = function() {{ \n',
            '     paste0("{prefix}_gene_sig_", Sys.Date(), ".png") \n',
            '   }}, \n',
            '   content = function(file) {{ ggsave( \n',
            '     file, device = "png", height = input$gsig_download_height, width = input$gsig_download_width, \n',
            '     plot = scGeneSig(input$gsig_list, input$gsig_group, input$gsig_subset, input$gsig_subset2, \n',
            '         input$gsig_subset_toggle, input$gsig_graphics_point_size, input$gsig_graphics_font_size, \n',
            '         input$gsig_graphics_color, input$gsig_graphics_cell_info) \n',
            '     ) \n',
            '   }} \n',
            ' ) \n',
            ' \n'
    )
  }
  else {
    return('')
  }
  #return(graphs)
}

#' Insert data server cues for differential gene expression volcano plot tab
#'
#' @param prefix file prefix
#' @param volc.plot TRUE/FALSE whether to create this tab
#' @rdname wrSVvolc
#' @export wrSVvolc
#'
wrSVvolc <- function(prefix, volc.plot) {
  #graphs <- ''
  if(volc.plot == TRUE) {
  #  graphs <- graphs + glue::glue(
  glue::glue(
            '  #  diff exp volcano plot \n',
            '  scVolc <- function(conf, genes, de_selection, genes_selection, de_subset_col, de_subset, fold_change, p_value, subset_toggle, \n',
            '                     top10_toggle, xlim_down, xlim_up, ylim, down_col, up_col, not_sig_color) {{ \n',
            '    \n',
            '    genes <- subset(genes, de_name == de_selection) \n',
            '    \n',
            '    if(tryCatch(!is.matrix(col2rgb(down_col)), error = function(e) TRUE)) {{ \n',
            '      down_col <- "#00798c" \n',
            '    }} \n',
            '    \n', 
            '    if(tryCatch(!is.matrix(col2rgb(up_col)), error = function(e) TRUE)) {{ \n',
            '      up_col <- "#d1495b" \n',
            '    }} \n',
            '    \n',
            '    if(tryCatch(!is.matrix(col2rgb(not_sig_color)), error = function(e) TRUE)) {{ \n',
            '      not_sig_color <- "grey82" \n',
            '    }} \n',
            '    \n',
            '    p_value <- as.numeric(p_value) \n',
            '    if(is.na(p_value)) {{ \n', 
            '      p_value <- 0.05 \n',
            '    }} \n', 
            '    \n',  
            '    #if(subset_toggle == 0) {{ # shotty hotfix for blank generated graph when subset menu hasn\'t been opened yet... \n',
            '    #  de_genes_subset <- unique(de_genes$cell_type) \n',
            '    #}} \n',
            '    \n', 
            '    if(subset_toggle == 0) {{ \n',
            '      orig_subset <- strsplit(conf$DEs[3], "\\\\|")[[1]][1] \n',
            '      #orig_subset <- de_subset_col \n',
            '      de_subset <- unique(genes[[orig_subset]]) \n',
            '    }} \n',
            '    \n', 
            '    #genes <- subset(genes, cell_type %in% de_subset) \n', 
            '    genes <- dplyr::filter(genes, !!as.name(de_subset_col) %in% de_subset) \n',
            '    genes$set_size <- 2.0 \n',
            '    \n',
            '    genes_selection_subset<-NULL \n',
            '    if(!is.null(genes_selection)) {{ \n',
            '      genes_selection <- unique(trimws(strsplit(genes_selection, ",|;| \n',
            '                                    |\\\\n")[[1]])) \n',
            '      #print(genes_selection) \n',
            '      #print("--------") \n',
            '      genes_selection_subset <- subset(genes, genes %in% genes_selection) \n',
            '      #print(paste0(genes_selection_subset)) \n',
            '      #print("---------") \n',
            '    }} \n',
            '    \n',
            '    if(top10_toggle == TRUE) {{ \n',
            '      top_pos <- subset(genes, log2FoldChange >= fold_change) \n',
            '      top_pos <- subset(top_pos, pvalue <= p_value) \n',
            '      top_pos <- top_pos[ order(top_pos$log2FoldChange, decreasing=TRUE)[1:10], ] \n',
            '      top_neg <- subset(genes, log2FoldChange <= -fold_change) \n',
            '      top_neg <- subset(top_neg, pvalue <= p_value) \n',
            '      top_neg <- top_neg[ order(top_neg$log2FoldChange)[1:10], ] \n',
            '      top_selection <- rbind(top_pos, top_neg) \n',
            '      genes_selection_subset <- rbind(genes_selection_subset, top_selection) \n',
            '    }} \n',
            '    \n',
            '    #print(genes_selection_subset) \n',
            '    \n',
            '    ggOut <- ggvolc(genes, genes_selection_subset, fc=fold_change, p_value=p_value, size_var="set_size", not_sig_color=not_sig_color, down_reg_color=down_col, up_reg_color=up_col) + labs(title=de_selection) + xlim(xlim_down, xlim_up) + ylim(0, ylim) \n',
            '    \n',
            '    return(ggOut) \n',
            '  }} \n',  
            '  \n',
            '  output$volc_subset.ui <- renderUI({{ \n', 
            '    subset = unique({prefix}de_genes_ggvolc[[input$volc_subset]]) # should you find a way to encode this into the {prefix}conf file?\n',
            '    checkboxGroupInput("volc_subset2", "Select which subsets to show", inline = TRUE, \n',
            '                       choices = subset, selected = subset) \n',
            '  }}) \n', 
            '  \n',
            '  observeEvent(input$volc_subset_sel_all, {{ \n', 
            '    subset = unique({prefix}de_genes_ggvolc[[input$volc_subset]]) \n',
            '    updateCheckboxGroupInput(session, inputId = "volc_subset2", label = "Select which subsets to show", \n',
            '                             choices = subset, selected = subset, inline = TRUE) \n',
            '  }}) \n', 
            '  \n',
            '  observeEvent(input$volc_subset_desel_all, {{ \n', 
            '    subset = unique({prefix}de_genes_ggvolc[[input$volc_subset]]) \n',
            '    updateCheckboxGroupInput(session, inputId = "volc_subset2", label = "Select which subsets to show", \n',
            '                             choices = subset, selected = NULL, inline = TRUE) \n',
            '  }}) \n',
            '  \n',
            '  output$volc_plot <- renderPlot({{ \n',
            '    # if seurat@misc$DE_genes$libra$overall was NULL upon site creation \n',
            '    if({prefix}conf$DEs[2] == "ERROR" & {prefix}conf$DEs[3] == "ERROR") {{ \n',
            '      ggplot() + labs(title="ERROR: a new \'{prefix}de_genes_ggvolc.rds\' was not created for this ShinyCellPlus instance!") \n',
            '    }} \n',
            '    # if {prefix}de_genes_ggvolc.rds not found \n',
            '    else if({prefix}de_genes_ggvolc[1,1] == "ERROR") {{ \n', 
            '      ggplot() + labs(title="ERROR: data file \'{prefix}de_genes_ggvolc.rds\' not found, cannot create plots!") \n',
            '    }} \n', 
            '    else {{ \n', 
            '      scVolc({prefix}conf, {prefix}de_genes_ggvolc, input$volc_de_select, genes_selection=input$volc_gene_highlight, input$volc_subset, input$volc_subset2, \n',
            '             input$volc_fold_change, input$volc_p_value, input$volc_subset_toggle, input$volc_top10_toggle, input$volc_xlim_down, input$volc_xlim_up, input$volc_ylim, input$volc_down_col, input$volc_up_col, input$volc_not_sig_col) \n',
            '    }} \n',
            '  }}) \n',
            '  output$volc_download.pdf <- downloadHandler( \n',
            '    filename = function() {{ \n',
            '      paste0("{prefix}_DE_", input$volc_de_select, "_volc_plot_", Sys.Date(), ".pdf") \n',
            '    }}, \n',
            '    content = function(file) {{ ggsave( \n',
            '      file, device = "pdf", height = input$volc_download_height, width = input$volc_download_width, useDingbats = FALSE,\n',
            '      plot = scVolc({prefix}conf, {prefix}de_genes_ggvolc, input$volc_de_select, genes_selection=input$volc_gene_highlight, input$volc_subset, input$volc_subset2, \n',
            '             input$volc_fold_change, input$volc_p_value, input$volc_subset_toggle, input$volc_top10_toggle, input$volc_xlim_down, input$volc_xlim_up, input$volc_ylim, input$volc_down_col, input$volc_up_col, input$volc_not_sig_col) \n',
            '      ) \n',
            '    }} \n',
            '  ) \n',
            '  \n',
            '  output$volc_download.png <- downloadHandler( \n',
            '    filename = function() {{ \n',
            '      paste0("{prefix}_DE_", input$volc_de_select, "_volc_plot_", Sys.Date(), ".png") \n',
            '    }}, \n',
            '    content = function(file) {{ ggsave( \n',
            '      file, device = "png", height = input$volc_download_height, width = input$volc_download_width, \n',
            '      plot = scVolc({prefix}conf, {prefix}de_genes_ggvolc, input$volc_de_select, genes_selection=input$volc_gene_highlight, input$volc_subset, input$volc_subset2, \n',
            '             input$volc_fold_change, input$volc_p_value, input$volc_subset_toggle, input$volc_top10_toggle, input$volc_xlim_down, input$volc_xlim_up, input$volc_ylim, input$volc_down_col, input$volc_up_col, input$volc_not_sig_col) \n',
            '      ) \n',
            '    }} \n',
            '  ) \n',
            .trim=FALSE
  )
    #)
  }
  else {
    return('')
  }
  #return(graphs)
}

#' Insert data server cues for ToppGene ontology tab
#'
#' @param prefix file prefix
#' @param gene.ont TRUE/FALSE whether to create this tab
#' @rdname wrSVgeneOnt
#' @export wrSVgeneOnt
#'
wrSVgeneOnt <- function(prefix, gene.ont) {
  if(gene.ont == TRUE) {
    glue::glue(
              '  \n',
              '  # ToppGene ontology \n',
              '  scTopp <- function(gene_ont, de_select, plot_select, category_select, balloons_select, clusters_select, text_size, show_empty) {{ \n',
              '    gene_ont <- filter(gene_ont, de_name == de_select) \n',
              '    # add more data prep here to cut down on all of the following conditionals for legibility? \n',
              '    \n',
              '    if(plot_select == "Balloon plot") {{ \n', 
              '      toppBalloon(gene_ont, categories=category_select, balloons=balloons_select, x_axis_text_size=text_size) \n',
              '    }} \n', 
              '    else {{ \n', 
              '      p <- toppPlot(gene_ont, category=category_select, clusters=clusters_select, y_axis_text_size=text_size) \n',
              '      \n',
              '      # special cases (no clusters selected, clusters selected with no data)... \n',
              '      if(length(clusters_select) <= 0 ) {{ # if no clusters are selected... \n',
              '        return(ggplot() + labs(title="No clusters selected.")) \n',
              '      }} \n',
              '      else if(length(clusters_select) == 1) {{ # if one cluster is selected... (structure of gene_ont object is different when containing 1 cluster) \n',
              '        if(nrow(p$data) <= 0) {{ # ...and there is no plot data in that cluster... \n',
              '          if(show_empty == TRUE) {{ \n',
              '            return(ggplot() + labs(title=paste0(category_select, ", Cluster ", clusters_select, "\\n(No ToppGene category data)"))) \n',
              '          }} \n',
              '          else {{ \n',
              '            cat("\\nNo data in cluster", clusters_select, ", removing it from graph object...\\n") \n',
              '            return(ggplot() + labs(title="Differential expression query to ToppGene did not return data relevant to this cluster.")) \n',
              '          }} \n',
              '        }} \n',
              '      }} \n',
              '      else {{ # if multiple clusters are selected... \n',
              '        for(i in clusters_select) {{ \n',
              '          if(nrow(p[[i]]$data) <= 0 ) {{ # if a selected cluster has no plot data... \n',
              '            if(show_empty == TRUE) {{ \n',
              '              p[[i]] <- ggplot() + labs(title=paste0("Cluster ", i, "\\n(No ToppGene category data)")) \n',
              '            }} \n',
              '            else {{ \n',
              '              cat("\\nNo data in cluster", as.character(i), ", removing it from graph object...\\n") \n',
              '              clusters_select <- clusters_select[!clusters_select %in% i] \n',
              '              p[[i]] <- NULL \n',
              '            }} \n',
              '          }} \n',
              '        }} \n',
              '        if(length(clusters_select) <= 0) {{ # if all of selected clusters have no data... \n',
              '          return(ggplot() + labs(title="Differential expression queries to ToppGene did not return data relevant to these clusters.")) \n',
              '        }} \n',
              '      }} \n',
              '      \n',
              '      # general expected case (selected clusters have plot data) \n',
              '      if(length(clusters_select) > 1) {{ # if multiple clusters were selected and multiple remain from removal (a list of ggplots is returned)... \n', 
              '        gg <- ggplot() \n',
              '        first <- TRUE \n',
              '        \n',
              '        for(i in clusters_select) {{ \n', 
              '          if(first == TRUE) {{ \n',
              '            if(length(p[[i]]$data) <= 0) {{ \n',
              '              gg <- p[[i]] + labs(title=paste0(category_select, ", Cluster ", i, "\\n(No ToppGene category data)")) \n',
              '            }} \n',
              '            else {{ \n',
              '              gg <- p[[i]] + labs(title=paste0(category_select, ", Cluster ", i)) \n',
              '            }} \n',
              '            first <- FALSE \n',
              '            next \n',
              '          }} \n', 
              '          gg <- gg + p[[i]] \n',
              '        }} \n',
              '        \n',
              '        return(gg) \n',
              '      }} \n',
              '      else {{ \n',
              '        if(class(p)[1] != "gg") {{ # if multiple clusters were selected and only one of them wasn\'t empty (a list of ggplots of length 1 is returned)...  \n',
              '          return(p[[clusters_select]] + labs(title=paste0(category_select, ", Cluster ", clusters_select))) \n',
              '        }} \n',
              '        else {{ # if only one cluster was selected and it has data... \n',
              '          return(p + labs(title=paste0(category_select, ", Cluster ", clusters_select))) \n',
              '        }} \n',
              '      }} \n',
              '    }} \n',
              '  }} \n',
              '  \n',
              '  # plotSizeFixer <- function(l_or_w) {{ \n',
              '  #   if(str_sub(l_or_w, start=length(l_or_w)-3) == "px") {{ \n',
              '  #     l_or_w <- strsplit(l_or_w, split="px")[[1]] \n',
              '  #   }} \n',
              '  #   else if(is.na(as.integer(l_or_w))) {{ \n',
              '  #     return(800) \n',
              '  #   }} \n',
              '  #   return(as.integer(l_or_w)) \n',
              '  # }} \n',
              '  \n',
              '  output$gont_cat_select.ui <- renderUI({{ \n',
              '    cats = get_ToppCats() \n',
              '    selectInput("gont_cat_select2", "Select category:", choices = cats) \n',
              '  }}) \n', 
              '  \n', 
              '  output$gont_cluster_vals.ui <- renderUI({{ \n',
              '    clusters = unique({prefix}gene_ont$Cluster) \n',
              '    checkboxGroupInput("gont_cluster_vals2", "Select clusters:", inline = TRUE, \n',
              '                       choices = as.character(sort(as.integer(clusters))), selected = clusters[1]) \n',
              '  }}) \n',
              '  \n',
              '  observeEvent(input$gont_cluster_vals_sel_all, {{ \n', 
              '    clusters = unique({prefix}gene_ont$Cluster) \n',
              '    updateCheckboxGroupInput(session, inputId = "gont_cluster_vals2", label = "Select clusters:", \n',
              '                             choices = as.character(sort(as.integer(clusters))), selected = clusters, inline = TRUE) \n',
              '  }}) \n',
              '  \n',
              '  observeEvent(input$gont_cluster_vals_desel_all, {{ \n', 
              '    clusters = unique({prefix}gene_ont$Cluster) \n',
              '    updateCheckboxGroupInput(session, inputId = "gont_cluster_vals2", label = "Select clusters:", \n',
              '                             choices = as.character(sort(as.integer(clusters))), selected = NULL, inline = TRUE) \n',
              '  }}) \n',
              '  \n', 
              '  output$gont_plot.ui <- renderUI({{ \n',
              '    plotOutput("gont_plot", width=input$gont_plot_x, height=input$gont_plot_y) \n',
              '  }}) \n',
              '  \n',
              '  output$gont_plot <- renderPlot({{ \n', 
              '    # if seurat@misc$DE_genes$libra$overall was NULL upon site creation \n',
              '    if({prefix}conf$DEs[4] == "ERROR") {{ \n',
              '      ggplot() + labs(title="ERROR: a new \'{prefix}gene_ont.rds\' was not created for this ShinyCellPlus instance!") \n',
              '    }} \n',
              '    # if {prefix}gene_ont.rds not found \n',
              '    else if({prefix}gene_ont[1,1] == "ERROR") {{ \n', 
              '      ggplot() + labs(title="ERROR: data file \'{prefix}gene_ont.rds\' not found, cannot create plots!") \n',
              '    }} \n',
              '    else {{ \n',
              '      scTopp({prefix}gene_ont, input$gont_de_select, input$gont_plot_select, input$gont_cat_select2, input$gont_balloon_val, input$gont_cluster_vals2, input$gont_text_size, input$gont_show_empty) \n',
              '    }} \n', 
              '  }}) \n',
              '  \n',      
              '  output$gont_download.pdf <- downloadHandler( \n',
              '    filename = function() {{ \n', 
              '      paste0("{prefix}_gene_ont_", input$gont_cat_select2, "_dot_plot_", Sys.Date(), ".pdf") \n',
              '    }}, \n', 
              '    content = function(file) {{ ggsave( \n',
              '      #file, device = "pdf", height = plotSizeFixer(input$gont_plot_y), width = 1.25*plotSizeFixer(input$gont_plot_x), scale=0.25, units="mm", useDingbats = FALSE, \n',
              '      file, device = "pdf", height = input$gont_download_height, width = input$gont_download_width, useDingbats = FALSE, \n',
              '      plot = scTopp({prefix}gene_ont, input$gont_de_select, input$gont_plot_select, input$gont_cat_select2, input$gont_balloon_val, input$gont_cluster_vals2, input$gont_text_size, input$gont_show_empty) \n',
              '    ) \n', 
              '    }} \n', 
              '  ) \n', 
              '  \n',
              '  output$gont_download.png <- downloadHandler( \n',
              '    filename = function() {{ \n', 
              '      paste0("{prefix}_gene_ont_", input$gont_cat_select2, "_dot_plot_", Sys.Date(), ".png") \n',
              '    }}, \n', 
              '    content = function(file) {{ ggsave( \n',
              '      #file, device = "png", height = plotSizeFixer(input$gont_plot_y), width = 1.25*plotSizeFixer(input$gont_plot_x), scale=0.25, units="mm", \n',
              '      file, device = "png", height = input$gont_download_height, width = input$gont_download_width, \n',
              '      plot = scTopp({prefix}gene_ont, input$gont_de_select, input$gont_plot_select, input$gont_cat_select2, input$gont_balloon_val, input$gont_cluster_vals2, input$gont_text_size, input$gont_show_empty) \n',
              '    ) \n', 
              '    }} \n', 
              '  ) \n', 
              .trim=FALSE
    )
  }
  else {
    return('')
  }
}

#' Write code for final portion of server.R
#'
#' @rdname wrSVend
#' @export wrSVend
#'
wrSVend <- function() {
  glue::glue('   \n',
             '   \n',
             '}}) \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n')
}

#' Write code for loading objects for ui.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrUIload
#' @export wrUIload
#'
wrUIload <- function(prefix) {
  glue::glue('{prefix}conf = readRDS("{prefix}conf.rds")\n',
             '{prefix}def  = readRDS("{prefix}def.rds")\n',
             '\n\n\n\n')
}

#' Write code for front portion of ui.R
#'
#' @param title shiny app title
#' @param ganalytics Google analytics tracking ID (e.g. "UA-123456789-0")
#'
#' @rdname wrUIsingle
#' @export wrUIsingle
#'
wrUIsingle <- function(title, ganalytics) {
  if(!is.na(ganalytics)){
    ga = 'tags$head(includeHTML(("google-analytics.html"))),'
  } else {
    ga = ''
  }
  glue::glue('### Start server code \n',
             'shinyUI(fluidPage( \n',
             '### HTML formatting of error messages \n',
             '{ga} \n',
             'tags$head(tags$style(HTML(".shiny-output-error-validation {{color: red; font-weight: bold;}}"))), \n',
             'list(tags$style(HTML(".navbar-default .navbar-nav {{ font-weight: bold; font-size: 16px; }}"))), \n',
             ' \n',
             '   \n',
             '### Page title \n',
             'titlePanel("{title}"),  \n',
             'navbarPage( \n',
             '  NULL,  \n',
             ' \n')
}

#' Write code for main block of ui.R
#'
#' @param prefix file prefix 
#'
#' @rdname wrUImain
#' @export wrUImain
#'
wrUImain <- function(prefix, subst = "", ptsiz = "1.25") {
  glue::glue('  ### Tab1.a1: cellInfo vs geneExpr on dimRed \n',
             '  tabPanel( \n',
             '    HTML("CellInfo vs GeneExpr"), \n',
             '    h4("Cell information vs gene expression on reduced dimensions"), \n',
             '    "In this tab, users can visualise both cell information and gene ",  \n',
             '    "expression side-by-side on low-dimensional representations.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        3, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            12, selectInput("{prefix}a1drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a1drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        3{subst}, actionButton("{prefix}a1togL", "Toggle to subset cells"), \n',
             '{subst}        conditionalPanel( \n',
             '{subst}          condition = "input.{prefix}a1togL % 2 == 1", \n',
             '{subst}          selectInput("{prefix}a1sub1", "Cell information to subset:", \n',
             '{subst}                      choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                      selected = {prefix}def$grp1), \n',
             '{subst}          uiOutput("{prefix}a1sub1.ui"), \n',
             '{subst}          actionButton("{prefix}a1sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}          actionButton("{prefix}a1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a1tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a1tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, sliderInput("{prefix}a1siz", "Point size:", \n',
             '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
             '              radioButtons("{prefix}a1psz", "Plot size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a1fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a1asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a1txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Cell information"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a1inp1", "Cell information:", \n',
             '                           choices = {prefix}conf$UI[!is.na({prefix}conf$UI)], \n',
             '                           selected = {prefix}def$meta1) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a1tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a1tog1 % 2 == 1", \n',
             '              radioButtons("{prefix}a1col1", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a1ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Original", inline = TRUE), \n',
             '              checkboxInput("{prefix}a1lab1", "Show cell info labels", value = TRUE), \n',
             '              checkboxInput("{prefix}a1legend", "Show legend", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a1oup1.ui"))), \n',
             '        downloadButton("{prefix}a1oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a1oup1.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a1oup1.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a1oup1.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)), br(), \n',
             '        actionButton("{prefix}a1tog9", "Toggle to show cell numbers / statistics"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a1tog9 % 2 == 1", \n',
             '          h4("Cell numbers / statistics"), \n',
             '          radioButtons("{prefix}a1splt", "Split continuous cell info into:", \n',
             '                       choices = c("Quartile", "Decile"), \n',
             '                       selected = "Decile", inline = TRUE), \n',
             '          dataTableOutput("{prefix}a1.dt") \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Gene expression"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a1inp2", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a1tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a1tog2 % 2 == 1", \n',
             '              radioButtons("{prefix}a1col2", "Colour:", \n',
             '                           choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a1ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ) , \n',
             '        fluidRow(column(12, uiOutput("{prefix}a1oup2.ui"))), \n',
             '        downloadButton("{prefix}a1oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a1oup2.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a1oup2.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a1oup2.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)) \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             '  ### Tab1.a2: cellInfo vs cellInfo on dimRed \n',
             '  tabPanel( \n',
             '    HTML("CellInfo vs CellInfo"), \n',
             '    h4("Cell information vs cell information on dimension reduction"), \n',
             '    "In this tab, users can visualise two cell informations side-by-side ", \n',
             '    "on low-dimensional representations.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        3, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            12, selectInput("{prefix}a2drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a2drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        3{subst}, actionButton("{prefix}a2togL", "Toggle to subset cells"), \n',
             '{subst}        conditionalPanel( \n',
             '{subst}          condition = "input.{prefix}a2togL % 2 == 1", \n',
             '{subst}          selectInput("{prefix}a2sub1", "Cell information to subset:", \n',
             '{subst}                      choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                      selected = {prefix}def$grp1), \n',
             '{subst}          uiOutput("{prefix}a2sub1.ui"), \n',
             '{subst}          actionButton("{prefix}a2sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}          actionButton("{prefix}a2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a2tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a2tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, sliderInput("{prefix}a2siz", "Point size:", \n',
             '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
             '              radioButtons("{prefix}a2psz", "Plot size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a2fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a2asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a2txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Cell information 1"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a2inp1", "Cell information:", \n',
             '                           choices = {prefix}conf$UI[!is.na({prefix}conf$UI)], \n',
             '                           selected = {prefix}def$meta1) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a2tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a2tog1 % 2 == 1", \n',
             '              radioButtons("{prefix}a2col1", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a2ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Original", inline = TRUE), \n',
             '              checkboxInput("{prefix}a2lab1", "Show cell info labels", value = TRUE), \n',
             '              checkboxInput("{prefix}a2legend1", "Show legend", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a2oup1.ui"))), \n',
             '        downloadButton("{prefix}a2oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a2oup1.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a2oup1.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a2oup1.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Cell information 2"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a2inp2", "Cell information:", \n',
             '                           choices = {prefix}conf$UI[!is.na({prefix}conf$UI)], \n',
             '                           selected = {prefix}def$meta2) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Cell information to colour cells by", \n',
             '                     content = c("Select cell information to colour cells", \n',
             '                                 "- Categorical covariates have a fixed colour palette", \n',
             '                                 paste0("- Continuous covariates are coloured in a ",  \n',
             '                                        "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a2tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a2tog2 % 2 == 1", \n',
             '              radioButtons("{prefix}a2col2", "Colour (Continuous data):", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "Blue-Yellow-Red"), \n',
             '              radioButtons("{prefix}a2ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Original", inline = TRUE), \n',
             '              checkboxInput("{prefix}a2lab2", "Show cell info labels", value = TRUE), \n',
             '              checkboxInput("{prefix}a2legend2", "Show legend", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a2oup2.ui"))), \n',
             '        downloadButton("{prefix}a2oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a2oup2.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a2oup2.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a2oup2.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)) \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             '  ### Tab1.a3: geneExpr vs geneExpr on dimRed \n',
             '  tabPanel( \n',
             '    HTML("GeneExpr vs GeneExpr"), \n',
             '    h4("Gene expression vs gene expression on dimension reduction"), \n',
             '    "In this tab, users can visualise two gene expressions side-by-side ", \n',
             '    "on low-dimensional representations.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        3, h4("Dimension Reduction"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            12, selectInput("{prefix}a3drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '            selectInput("{prefix}a3drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                        selected = {prefix}def$dimred[2])) \n',
             '        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        3{subst}, actionButton("{prefix}a3togL", "Toggle to subset cells"), \n',
             '{subst}        conditionalPanel( \n',
             '{subst}          condition = "input.{prefix}a3togL % 2 == 1", \n',
             '{subst}          selectInput("{prefix}a3sub1", "Cell information to subset:", \n',
             '{subst}                      choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                      selected = {prefix}def$grp1), \n',
             '{subst}          uiOutput("{prefix}a3sub1.ui"), \n',
             '{subst}          actionButton("{prefix}a3sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}          actionButton("{prefix}a3sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}        ) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, actionButton("{prefix}a3tog0", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}a3tog0 % 2 == 1", \n',
             '          fluidRow( \n',
             '            column( \n',
             '              6, sliderInput("{prefix}a3siz", "Point size:", \n',
             '                             min = 0, max = 4, value = {ptsiz}, step = 0.25), \n',
             '              radioButtons("{prefix}a3psz", "Plot size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE), \n',
             '              radioButtons("{prefix}a3fsz", "Font size:", \n',
             '                           choices = c("Small", "Medium", "Large"), \n',
             '                           selected = "Medium", inline = TRUE) \n',
             '            ), \n',
             '            column( \n',
             '              6, radioButtons("{prefix}a3asp", "Aspect ratio:", \n',
             '                              choices = c("Square", "Fixed", "Free"), \n',
             '                              selected = "Square", inline = TRUE), \n',
             '              checkboxInput("{prefix}a3txt", "Show axis text", value = FALSE) \n',
             '            ) \n',
             '          ) \n',
             '        ) \n',
             '      )  # End of column (6 space) \n',
             '    ),   # End of fluidRow (4 space) \n',
             '    fluidRow( \n',
             '      column( \n',
             '        6, style="border-right: 2px solid black", h4("Gene expression 1"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a3inp1", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a3tog1", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a3tog1 % 2 == 1", \n',
             '              radioButtons("{prefix}a3col1", "Colour:", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a3ord1", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a3oup1.ui"))), \n',
             '        downloadButton("{prefix}a3oup1.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a3oup1.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a3oup1.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a3oup1.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)) \n',
             '      ), # End of column (6 space) \n',
             '      column( \n',
             '        6, h4("Gene expression 2"), \n',
             '        fluidRow( \n',
             '          column( \n',
             '            6, selectInput("{prefix}a3inp2", "Gene name:", choices=NULL) %>%  \n',
             '              helper(type = "inline", size = "m", fade = TRUE, \n',
             '                     title = "Gene expression to colour cells by", \n',
             '                     content = c("Select gene to colour cells by gene expression", \n',
             '                                 paste0("- Gene expression are coloured in a ", \n',
             '                                        "White-Red colour scheme which can be ", \n',
             '                                        "changed in the plot controls"))) \n',
             '          ), \n',
             '          column( \n',
             '            6, actionButton("{prefix}a3tog2", "Toggle plot controls"), \n',
             '            conditionalPanel( \n',
             '              condition = "input.{prefix}a3tog2 % 2 == 1", \n',
             '              radioButtons("{prefix}a3col2", "Colour:", \n',
             '                           choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                       "Yellow-Green-Purple"), \n',
             '                           selected = "White-Red"), \n',
             '              radioButtons("{prefix}a3ord2", "Plot order:", \n',
             '                           choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                           selected = "Max-1st", inline = TRUE) \n',
             '            ) \n',
             '          ) \n',
             '        ), \n',
             '        fluidRow(column(12, uiOutput("{prefix}a3oup2.ui"))), \n',
             '        downloadButton("{prefix}a3oup2.pdf", "Download PDF"), \n',
             '        downloadButton("{prefix}a3oup2.png", "Download PNG"), br(), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a3oup2.h", "PDF / PNG height:", width = "138px", \n',
             '                         min = 4, max = 20, value = 6, step = 0.5)), \n',
             '        div(style="display:inline-block", \n',
             '            numericInput("{prefix}a3oup2.w", "PDF / PNG width:", width = "138px", \n',
             '                         min = 4, max = 20, value = 8, step = 0.5)) \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  ),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             ' ### Tab1.b2: Gene coexpression plot \n',
             ' tabPanel( \n',
             '    HTML("Gene Dual Coex."), \n',
             '    h4("Coexpression of two genes on reduced dimensions"), \n',
             '    "In this tab, users can visualise the coexpression of two genes ", \n',
             '    "on low-dimensional representations.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '    column( \n',
             '      3, h4("Dimension Reduction"), \n',
             '          selectInput("{prefix}b2drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                      selected = {prefix}def$dimred[1]), \n',
             '          selectInput("{prefix}b2drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                      selected = {prefix}def$dimred[2]), \n',
             '          actionButton("{prefix}b2tog1", "Toggle plot controls"), \n',
             '          conditionalPanel( \n',
             '            condition = "input.{prefix}b2tog1 % 2 == 1", \n',
             '            radioButtons("{prefix}b2col1", "Colour:", \n',
             '                         choices = c("Red (Gene1); Blue (Gene2)", \n',
             '                                     "Orange (Gene1); Blue (Gene2)", \n',
             '                                     "Red (Gene1); Green (Gene2)", \n',
             '                                     "Green (Gene1); Blue (Gene2)"), \n',
             '                         selected = "Red (Gene1); Blue (Gene2)"), \n',
             '            radioButtons("{prefix}b2ord1", "Plot order:", \n',
             '                         choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                         selected = "Max-1st", inline = TRUE) \n',
             '    ), \n',
             '    br(),br(), \n',
             '    actionButton("{prefix}b2togL", "Toggle to subset cells"), \n',
             '    conditionalPanel( \n',
             '      condition = "input.{prefix}b2togL % 2 == 1", \n',
             '      selectInput("{prefix}b2sub1", "Cell information to subset:", \n',
             '                  choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                  selected = {prefix}def$grp1), \n',
             '      uiOutput("{prefix}b2sub1.ui"), \n',
             '      actionButton("{prefix}b2sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '      actionButton("{prefix}b2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '    ), \n',
             '    br(),br(), \n',
             '    actionButton("{prefix}b2tog0", "Toggle graphics controls"), \n',
             '    conditionalPanel( \n',
             '      condition = "input.{prefix}b2tog0 % 2 == 1", \n',
             '      sliderInput("{prefix}b2siz", "Point size:", \n',
             '                  min = 0, max = 4, value = 1.25, step = 0.25), \n',
             '      radioButtons("{prefix}b2psz", "Plot size:", \n',
             '                   choices = c("Small", "Medium", "Large"), \n',
             '                   selected = "Medium", inline = TRUE), \n',
             '      radioButtons("{prefix}b2fsz", "Font size:", \n',
             '                   choices = c("Small", "Medium", "Large"), \n',
             '                   selected = "Small", inline = TRUE), \n',
             '      radioButtons("{prefix}b2asp", "Aspect ratio:", \n',
             '                   choices = c("Square", "Fixed", "Free"), \n',
             '                   selected = "Square", inline = TRUE), \n',
             '      checkboxInput("{prefix}b2txt", "Show axis text", value = FALSE) \n',
             '    ) \n',
             '  ), # End of column (6 space) \n',
             '  \n',
             '  column( \n',
             '    4,style="border-left: 2px solid black", h4("Gene Expression 1"), \n',
             '    selectInput("{prefix}b2inp1", "Gene 1:", choices=NULL) %>%  \n',
             '      helper(type = "inline", size = "m", fade = TRUE, \n',
             '             title = "Gene expression to colour cells by", \n',
             '             content = c("Select gene to colour cells by gene expression", \n',
             '                         paste0("- Gene expression are coloured in a ", \n',
             '                                "White-Red colour scheme which can be ", \n',
             '                                "changed in the plot controls"))), \n',
             '    selectInput("{prefix}b2inp2", "Gene 2:", choices=NULL) %>% \n',
             '      helper(type = "inline", size = "m", fade = TRUE, \n',
             '             title = "Gene expression to colour cells by", \n',
             '             content = c("Select gene to colour cells by gene expression", \n',
             '                         paste0("- Gene expression are coloured in a ", \n',
             '                                "White-Blue colour scheme which can be ", \n',
             '                                "changed in the plot controls"))), \n',
             '    #style="border-right: 2px solid black", \n',
             '    uiOutput("{prefix}b2oup1.ui"), \n',
             '    downloadButton("{prefix}b2oup1.pdf", "Download PDF"), \n',
             '    downloadButton("{prefix}b2oup1.png", "Download PNG"), br(), \n',
             '    div(style="display:inline-block", \n',
             '        numericInput("{prefix}b2oup1.h", "PDF / PNG height:", width = "138px", \n',
             '                     min = 4, max = 20, value = 8, step = 0.5)), \n',
             '    div(style="display:inline-block", \n',
             '    numericInput("{prefix}b2oup1.w", "PDF / PNG width:", width = "138px", \n',
             '                 min = 4, max = 20, value = 10, step = 0.5)), \n',
             '    uiOutput("{prefix}b2oup2.ui"), \n',
             '    downloadButton("{prefix}b2oup2.pdf", "Download PDF"), \n',
             '    downloadButton("{prefix}b2oup2.png", "Download PNG"), \n',
             '    br(), h4("Cell numbers"), \n',
             '    dataTableOutput("{prefix}b2.dt") \n',
             '  ), # End of column (6 space) \n',
             '  \n',
             '  column( \n',
             '    4, style="border-left: 2px solid black", h4("Gene Expression 2"), \n',
             '    selectInput("{prefix}b2inp3", "Gene 1:", choices=NULL) %>% \n',
             '      helper(type = "inline", size = "m", fade = TRUE, \n',
             '             title = "Gene expression to colour cells by", \n',
             '             content = c("Select gene to colour cells by gene expression", \n',
             '                         paste0("- Gene expression are coloured in a ", \n',
             '                         "White-Red colour scheme which can be ", \n',
             '                         "changed in the plot controls"))), \n',
             '    selectInput("{prefix}b2inp4", "Gene 2:", choices=NULL) %>% \n',
             '      helper(type = "inline", size = "m", fade = TRUE, \n',
             '             title = "Gene expression to colour cells by", \n',
             '             content = c("Select gene to colour cells by gene expression", \n',
             '                         paste0("- Gene expression are coloured in a ", \n',
             '                                "White-Blue colour scheme which can be ", \n',
             '                                "changed in the plot controls"))), \n',
             '    #style="border-right: 2px solid black", \n',
             '    uiOutput("{prefix}b2oup3.ui"), \n',
             '    downloadButton("{prefix}b2oup3.pdf", "Download PDF"), \n',
             '    downloadButton("{prefix}b2oup3.png", "Download PNG"), br(), \n',
             '    div(style="display:inline-block", \n',
             '        numericInput("{prefix}b2oup3.h", "PDF / PNG height:", width = "138px", \n',
             '                     min = 4, max = 20, value = 8, step = 0.5)), \n',
             '    div(style="display:inline-block", \n',
             '    numericInput("{prefix}b2oup3.w", "PDF / PNG width:", width = "138px", \n',
             '                 min = 4, max = 20, value = 10, step = 0.5)), \n',
             '    uiOutput("{prefix}b2oup4.ui"), \n',
             '    downloadButton("{prefix}b2oup4.pdf", "Download PDF"), \n',
             '    downloadButton("{prefix}b2oup4.png", "Download PNG"), \n',
             '    br(), h4("Cell numbers"), \n',
             '    dataTableOutput("{prefix}b2_2.dt") \n',
             '  ) \n',
             ' ) \n',
             ' ),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             '### Tab1.a1b: SPLIT CELL INFO \n',
             ' tabPanel( \n',
             '   HTML("Split Dataset: Cell Info"), \n',
             '   h4("Cell information vs gene expression on reduced dimensions"), \n',
             '      "In this tab, users can split comparisons between cell meta ", \n',
             '      "data side-by-side while choosing a second meta field two visualize with color gradient on low-dimensional representations.", \n',
             '   br(),br(), \n',
             '   fluidRow( \n',
             '     column( \n',
             '       3, h4("Dimension Reduction"), \n',
             '       fluidRow( \n',
             '         column( \n',
             '           12, \n',
             '           selectInput("{prefix}a1bdrX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                           selected = {prefix}def$dimred[1]), \n',
             '           selectInput("{prefix}a1bdrY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                       selected = {prefix}def$dimred[2])) \n',
             '         ) \n',
             '       ), # End of column (6 space) \n',
             '       column( \n',
             '         3, \n',
             '         actionButton("{prefix}a1btogL", "Toggle to subset cells"), \n',
             '         conditionalPanel( \n',
             '           condition = "input.{prefix}a1btogL % 2 == 1", \n',
             '           selectInput("{prefix}a1bsub1", "Cell information to subset:", \n',
             '                       choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                       selected = {prefix}def$grp1), \n',
             '           uiOutput("{prefix}a1bsub1.ui"), \n',
             '           actionButton("{prefix}a1bsub1all", "Select all groups", class = "btn btn-primary"), \n',
             '           actionButton("{prefix}a1bsub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '         ) \n',
             '       ), # End of column (6 space) \n',
             '       column( \n',
             '       6, \n',
             '       actionButton("{prefix}a1btog0", "Toggle graphics controls"), \n',
             '       conditionalPanel( \n',
             '         condition = "input.{prefix}a1btog0 % 2 == 1", \n',
             '         fluidRow( \n',
             '           column( \n',
             '             6, \n',
             '             sliderInput("{prefix}a1bsiz", "Point size:", \n',
             '                         min = 0, max = 4, value = 1.25, step = 0.25), \n',
             '             radioButtons("{prefix}a1bpsz", "Plot size:", \n',
             '                          choices = c("Small", "Medium", "Large"), \n',
             '                          selected = "Medium", inline = TRUE), \n',
             '             radioButtons("{prefix}a1bfsz", "Font size:", \n',
             '                          choices = c("Small", "Medium", "Large"), \n',
             '                          selected = "Medium", inline = TRUE) \n',
             '           ), \n',
             '           column( \n',
             '             6, \n',
             '             radioButtons("{prefix}a1basp", "Aspect ratio:", \n',
             '                          choices = c("Square", "Fixed", "Free"), \n',
             '                          selected = "Square", inline = TRUE), \n',
             '             checkboxInput("{prefix}a1btxt", "Show axis text", value = FALSE) \n',
             '           ) \n',
             '         ) \n',
             '       ) \n',
             '     )  # End of column (6 space) \n',
             '   ),   # End of fluidRow (4 space) \n',
             '   fluidRow( \n',
             '     column( \n',
             '       6, style="border-right: 2px solid black", \n',
             '       h4("Split dataset"), \n',
             '       fluidRow( \n',
             '         column( \n',
             '           12, \n',
             '           selectInput("{prefix}a1bsplit1", "Split dataset:", \n',
             '                       choices = {prefix}conf[{prefix}conf$split==TRUE]$UI), \n',
             '           uiOutput("{prefix}a1bsplit_select1_out1"), \n',
             '           uiOutput("{prefix}a1bsplit_select2_out1") \n',
             '         ) \n',
             '       ) \n',
             '     ) \n',
             '   ), \n',
             '   fluidRow( \n',
             '     column( \n',
             '       6, style="border-right: 2px solid black", \n',
             '       h4("Cell information"), \n',
             '       fluidRow( \n',
             '         column( \n',
             '           6, \n',
             '           selectInput("{prefix}a1binp1", "Cell information:", \n',
             '                       choices = {prefix}conf$UI[!is.na({prefix}conf$UI)], \n',
             '                       selected = {prefix}def$meta1) %>% \n',
             '                       helper(type = "inline", size = "m", fade = TRUE, \n',
             '                              title = "Cell information to colour cells by", \n',
             '                              content = c("Select cell information to colour cells", \n',
             '                                          "- Categorical covariates have a fixed colour palette", \n',
             '                                          paste0("- Continuous covariates are coloured in a ", \n',
             '                                          "Blue-Yellow-Red colour scheme, which can be ", \n',
             '                                          "changed in the plot controls"))) \n',
             '           ), \n',
             '           column( \n',
             '           6, \n',
             '           actionButton("{prefix}a1btog1", "Toggle plot controls"), \n',
             '           conditionalPanel( \n',
             '             condition = "input.{prefix}a1btog1 % 2 == 1", \n',
             '             radioButtons("{prefix}a1bcol1", "Colour (Continuous data):", \n',
             '                          choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '                          selected = "Blue-Yellow-Red"), \n',
             '             radioButtons("{prefix}a1bord1", "Plot order:", \n',
             '                          choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                          selected = "Original", inline = TRUE), \n',
             '             checkboxInput("{prefix}a1blab1", "Show cell info labels", value = TRUE), \n',
             '             checkboxInput("{prefix}a1blegend", "Show legend", value = FALSE) \n',
             '           ) \n',
             '         ) \n',
             '       ), \n',
             '       fluidRow( \n',
             '         column( \n',
             '           12, \n',
             '           uiOutput("{prefix}a1boup1.ui") \n',
             '         ) \n',
             '       ), \n',
             '       downloadButton("{prefix}a1boup1.pdf", "Download PDF"), \n',
             '       downloadButton("{prefix}a1boup1.png", "Download PNG"), br(), \n',
             '       div(style="display:inline-block", \n',
             '         numericInput("{prefix}a1boup1.h", "PDF / PNG height:", width = "138px", \n',
             '                      min = 4, max = 20, value = 6, step = 0.5)), \n',
             '       div(style="display:inline-block", \n',
             '         numericInput("{prefix}a1boup1.w", "PDF / PNG width:", width = "138px", \n',
             '                      min = 4, max = 20, value = 8, step = 0.5)), br(), \n',
             '#       actionButton("{prefix}a1btog9", "Toggle to show cell numbers / statistics"), \n',
             '#       conditionalPanel( \n',
             '#         condition = "input.{prefix}a1btog9 % 2 == 1", \n',
             '#         h4("Cell numbers / statistics"), \n',
             '#         radioButtons("{prefix}a1bsplt", "Split continuous cell info into:", \n',
             '#                      choices = c("Quartile", "Decile"), \n',
             '#                      selected = "Decile", inline = TRUE), \n',
             '#         dataTableOutput("{prefix}a1b.dt") \n',
             '#       ) \n',
             '     ), # End of column (6 space) \n',
             '     column( \n',
             '       6, \n',
             '       br(),br(), br(),br(),br(), br(), \n',
             '# fluidRow( \n',
             '#   column( \n',
             '#     6, selectInput("{prefix}a1binp1", "Cell information:", \n',
             '#                    choices = {prefix}conf$UI, \n',
             '#                    selected = {prefix}def$meta1) %>% \n',
             '#       helper(type = "inline", size = "m", fade = TRUE, \n',
             '#              title = "Cell information to colour cells by", \n',
             '#              content = c("Select cell information to colour cells", \n',
             '#                          "- Categorical covariates have a fixed colour palette", \n',
             '#                          paste0("- Continuous covariates are coloured in a ", \n',
             '#                                 "Blue-Yellow-Red colour scheme, which can be ", \n',
             '#                                 "changed in the plot controls"))) \n',
             '#   ), \n',
             '#   column( \n',
             '#     6, actionButton("{prefix}a1btog1", "Toggle plot controls"), \n',
             '#     conditionalPanel( \n',
             '#       condition = "input.{prefix}a1btog1 % 2 == 1", \n',
             '#       radioButtons("{prefix}a1bcol1", "Colour (Continuous data):", \n',
             '#                    choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '#                    selected = "Blue-Yellow-Red"), \n',
             '#       radioButtons("{prefix}a1bord1", "Plot order:", \n',
             '#                    choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '#                    selected = "Original", inline = TRUE), \n',
             '#       checkboxInput("{prefix}a1blab1", "Show cell info labels", value = TRUE) \n',
             '#     ) \n',
             '#   ) \n',
             '# ), \n',
             '# fluidRow( \n',
             '#   column( \n',
             '#     6, selectInput("{prefix}a1binp2", "Gene name:", choices=NULL) %>% \n',
             '#       helper(type = "inline", size = "m", fade = TRUE, \n',
             '#              title = "Gene expression to colour cells by", \n',
             '#              content = c("Select gene to colour cells by gene expression", \n',
             '#                          paste0("- Gene expression are coloured in a ", \n',
             '#                                 "White-Red colour scheme which can be ", \n',
             '#                                 "changed in the plot controls"))) \n',
             '#   ), \n',
             '#   column( \n',
             '#     6, actionButton("{prefix}a1btog2", "Toggle plot controls"), \n',
             '#     conditionalPanel( \n',
             '#       condition = "input.{prefix}a1tbog2 % 2 == 1", \n',
             '#       radioButtons("{prefix}a1cbol2", "Colour:", \n',
             '#                    choices = c("White-Red","Blue-Yellow-Red","Yellow-Green-Purple"), \n',
             '#                    selected = "White-Red"), \n',
             '#       radioButtons("{prefix}a1bord2", "Plot order:", \n',
             '#                    choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '#                    selected = "Max-1st", inline = TRUE) \n',
             '#     ) \n',
             '#   ) \n',
             '# ) , \n',
             '       fluidRow( \n',
             '         column( \n',
             '           12, \n',
             '           uiOutput("{prefix}a1boup2.ui") \n',
             '         ) \n',
             '       ), \n',
             '       downloadButton("{prefix}a1boup2.pdf", "Download PDF"), \n',
             '       downloadButton("{prefix}a1boup2.png", "Download PNG"), br(), \n',
             '       div(style="display:inline-block", \n',
             '         numericInput("{prefix}a1boup2.h", "PDF / PNG height:", width = "138px", \n',
             '                      min = 4, max = 20, value = 6, step = 0.5)), \n',
             '       div(style="display:inline-block", \n',
             '         numericInput("{prefix}a1boup2.w", "PDF / PNG width:", width = "138px", \n',
             '                      min = 4, max = 20, value = 8, step = 0.5)) \n',
             '     )  # End of column (6 space) \n',
             '   )    # End of fluidRow (4 space) \n',
             ' ),     # End of tab (2 space) \n',
             ' \n',
             '### Tab1.a1c: geneExpr vs geneExpr on dimRed \n',
             ' tabPanel( \n',
             '   HTML("Split Dataset: Gene Expression"), \n',
             '   h4("Gene expression vs gene expression on dimension reduction"), \n',
             '   "In this tab, users can visualise two gene expressions split by meta data side-by-side ", \n',
             '   "on low-dimensional representations.", \n',
             '   br(),br(), \n',
             '   fluidRow( \n',
             '     column( \n',
             '       2, \n',
             '       h4("Dimension Reduction"), \n',
             '       selectInput("{prefix}a1cdrX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                   selected = {prefix}def$dimred[1]), \n',
             '       selectInput("{prefix}a1cdrY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                   selected = {prefix}def$dimred[2]), \n',
             '       h4("Split dataset"), \n',
             '       selectInput("{prefix}a1csplit1", "Split dataset:", \n',
             '                   choices = {prefix}conf[{prefix}conf$split==TRUE]$UI), \n',
             '       uiOutput("{prefix}a1csplit_select1_out1"), \n',
             '       uiOutput("{prefix}a1csplit_select2_out1"), \n',
             '       h4("Gene expression"), \n',
             '       selectInput("{prefix}a1cinp1", "Gene name:", choices=NULL) %>% \n',
             '         helper(type = "inline", size = "m", fade = TRUE, \n',
             '                title = "Gene expression to colour cells by", \n',
             '                content = c("Select gene to colour cells by gene expression", \n',
             '                            paste0("- Gene expression are coloured in a ", \n',
             '                                   "White-Red colour scheme which can be ", \n',
             '                                   "changed in the plot controls"))), \n',
             '       actionButton("{prefix}a1ctogL", "Toggle to subset cells"), \n',
             '       conditionalPanel( \n',
             '         condition = "input.{prefix}a1ctogL % 2 == 1", \n',
             '         selectInput("{prefix}a1csub1", "Cell information to subset:", \n',
             '                     choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                     selected = {prefix}def$grp1), \n',
             '                     uiOutput("{prefix}a1csub1.ui"), \n',
             '         actionButton("{prefix}a1csub1all", "Select all groups", class = "btn btn-primary"), \n',
             '         actionButton("{prefix}a1csub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '       ), \n',
             '       br(),br(), \n',
             '       actionButton("{prefix}a1ctog0", "Toggle graphics controls"), \n',
             '       conditionalPanel( \n',
             '         condition = "input.{prefix}a1ctog0 % 2 == 1", \n',
             '         sliderInput("{prefix}a1csiz", "Point size:", \n',
             '                     min = 0, max = 4, value = 1.25, step = 0.25), \n',
             '         radioButtons("{prefix}a1cpsz", "Plot size:", \n',
             '                      choices = c("Small", "Medium", "Large"), \n',
             '                      selected = "Medium", inline = TRUE), \n',
             '         radioButtons("{prefix}a1cfsz", "Font size:", \n',
             '                      choices = c("Small", "Medium", "Large"), \n',
             '                      selected = "Medium", inline = TRUE), \n',
             '         radioButtons("{prefix}a1casp", "Aspect ratio:", \n',
             '                      choices = c("Square", "Fixed", "Free"), \n',
             '                      selected = "Square", inline = TRUE), \n',
             '         checkboxInput("{prefix}a1ctxt", "Show axis text", value = FALSE) \n',
             '       ), \n',
             '       br(),br(), \n',
             '       actionButton("{prefix}a1ctog1", "Toggle plot controls"), \n',
             '       conditionalPanel( \n',
             '         condition = "input.{prefix}a1ctog1 % 2 == 1", \n',
             '         radioButtons("{prefix}a1ccol1", "Colour:", \n',
             '                      choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                  "Yellow-Green-Purple"), \n',
             '                      selected = "White-Red"), \n',
             '         radioButtons("{prefix}a1cord1", "Plot order:", \n',
             '                      choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                      selected = "Max-1st", inline = TRUE) \n',
             '         ) \n', 
             '       ), \n',
             '       column( \n',
             '         5, \n',
             '         uiOutput("{prefix}a1coup1.ui"), \n',
             '         downloadButton("{prefix}a1coup1.pdf", "Download PDF"), \n',
             '         downloadButton("{prefix}a1coup1.png", "Download PNG"), \n',
             '         br(), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}a1coup1.h", "PDF / PNG height:", width = "138px", \n',
             '                          min = 4, max = 20, value = 6, step = 0.5) \n',
             '         ), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}a1coup1.w", "PDF / PNG width:", width = "138px", \n',
             '                          min = 4, max = 20, value = 8, step = 0.5) \n',
             '         ) \n',
             '       ), \n',
             '       column( \n',
             '         5, \n',
             '         uiOutput("{prefix}a1coup2.ui"), \n',
             '         downloadButton("{prefix}a1coup2.pdf", "Download PDF"), \n',
             '         downloadButton("{prefix}a1coup2.png", "Download PNG"), \n',
             '         br(), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}a1coup2.h", "PDF / PNG height:", width = "138px", \n',
             '                          min = 4, max = 20, value = 6, step = 0.5) \n',
             '         ), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}a1coup2.w", "PDF / PNG width:", width = "138px", \n',
             '                          min = 4, max = 20, value = 8, step = 0.5) \n',
             '         ) \n',
             '       ), \n',
             '     ) \n',
             '   ), \n',
             ' ### b3 \n',
             ' tabPanel( \n',
             '   HTML("Split Dataset: Gene Coexpression"), \n',
             '   h4("Coexpression of two genes on reduced dimensions"), \n',
             '   "In this tab, users can visualise the coexpression of two genes ", \n',
             '   "on low-dimensional representations, while splitting between meta data", \n',
             '   "columns which have more than two possibilities.", \n',
             '   br(),br(), \n',
             '   fluidRow( \n',
             '     column( \n',
             '       2, \n',
             '       h4("Dimension Reduction"), \n',
             '         selectInput("{prefix}b3drX", "X-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                     selected = {prefix}def$dimred[1]), \n',
             '         selectInput("{prefix}b3drY", "Y-axis:", choices = {prefix}conf[dimred == TRUE]$UI, \n',
             '                     selected = {prefix}def$dimred[2]), \n',
             '         h4("Gene Expression"), \n',
             '         selectInput("{prefix}b3inp1", "Gene 1:", choices=NULL) %>% \n',
             '           helper(type = "inline", size = "m", fade = TRUE, \n',
             '                  title = "Gene expression to colour cells by", \n',
             '                  content = c("Select gene to colour cells by gene expression", \n',
             '                              paste0("- Gene expression are coloured in a ", \n',
             '                                     "White-Red colour scheme which can be ", \n',
             '                                     "changed in the plot controls"))), \n',
             '         selectInput("{prefix}b3inp2", "Gene 2:", choices=NULL) %>% \n',
             '           helper(type = "inline", size = "m", fade = TRUE, \n',
             '                  title = "Gene expression to colour cells by", \n',
             '                  content = c("Select gene to colour cells by gene expression", \n',
             '                              paste0("- Gene expression are coloured in a ", \n',
             '                                     "White-Blue colour scheme which can be ", \n',
             '                                     "changed in the plot controls"))), \n',
             '         #selectInput("{prefix}b3spl", "Split by:", choices = NULL), \n',
             '         selectInput("{prefix}b3split", "Split by:", choices = {prefix}conf[{prefix}conf$split==TRUE]$UI), \n',
             '         uiOutput("{prefix}b3split_select1_out1"), \n',
             '         uiOutput("{prefix}b3split_select2_out1"), \n', 
             '         actionButton("{prefix}b3tog1", "Toggle plot controls"), \n',
             '         conditionalPanel( \n',
             '           condition = "input.{prefix}b3tog1 % 2 == 1", \n',
             '           radioButtons("{prefix}b3col1", "Colour:", \n',
             '                        choices = c("Red (Gene1); Blue (Gene2)", \n',
             '                                    "Orange (Gene1); Blue (Gene2)", \n',
             '                                    "Red (Gene1); Green (Gene2)", \n',
             '                                    "Green (Gene1); Blue (Gene2)"), \n',
             '                        selected = "Red (Gene1); Blue (Gene2)"), \n',
             '           radioButtons("{prefix}b3ord1", "Plot order:", \n',
             '                        choices = c("Max-1st", "Min-1st", "Original", "Random"), \n',
             '                        selected = "Max-1st", inline = TRUE) \n',
             '         ), \n',
             '         br(),br(), \n',
             '         actionButton("{prefix}b3togL", "Toggle to subset cells"), \n',
             '         conditionalPanel( \n',
             '           condition = "input.{prefix}b3togL % 2 == 1", \n',
             '           selectInput("{prefix}b3sub1", "Cell information to subset:", \n',
             '                       choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                       selected = {prefix}def$grp1), \n',
             '           uiOutput("{prefix}b3sub1.ui"), \n',
             '           actionButton("{prefix}b3sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '           actionButton("{prefix}b3sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '         ), \n',
             '         br(),br(), \n',
             '         actionButton("{prefix}b3tog0", "Toggle graphics controls"), \n',
             '         conditionalPanel( \n',
             '           condition = "input.{prefix}b3tog0 % 2 == 1", \n',
             '           sliderInput("{prefix}b3siz", "Point size:", \n',
             '                       min = 0, max = 4, value = 1.25, step = 0.25), \n',
             '           radioButtons("{prefix}b3psz", "Plot size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Medium", inline = TRUE), \n',
             '           radioButtons("{prefix}b3fsz", "Font size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Small", inline = TRUE), \n',
             '           radioButtons("{prefix}b3asp", "Aspect ratio:", \n',
             '                       choices = c("Square", "Fixed", "Free"), \n',
             '                       selected = "Square", inline = TRUE), \n',
             '           checkboxInput("{prefix}b3txt", "Show axis text", value = FALSE), \n',
             '           br(), br() \n',
             '         ), \n',
             '         column(12, \n',
             '           br(), br(), \n',
             '           uiOutput("{prefix}b3oup3.ui"), \n',
             '           downloadButton("{prefix}b3oup3.pdf", "Download PDF"), \n',
             '           downloadButton("{prefix}b3oup3.png", "Download PNG") \n',
             '         ) \n',
             '       ), # End of column (6 space) \n',
             '       \n',
             '       column( \n',
             '         5, \n',
             '         #style="border-right: 2px solid black", \n',
             '         uiOutput("{prefix}b3oup1.ui"), \n',
             '         downloadButton("{prefix}b3oup1.pdf", "Download PDF"), \n',
             '         downloadButton("{prefix}b3oup1.png", "Download PNG"), br(), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}b3oup1.h", "PDF / PNG height:", width = "138px", \n',
             '                          min = 4, max = 20, value = 8, step = 0.5)), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}b3oup1.w", "PDF / PNG width:", width = "138px", \n',
             '                          min = 4, max = 20, value = 10, step = 0.5)), \n',
             '         br(), h4("Cell numbers"), # paste0( plot title, " cell numbers") ? \n',
             '         dataTableOutput("{prefix}b3oup1.dt") \n',
             '       ), \n',
             '       column ( \n',
             '         5, \n',
             '         uiOutput("{prefix}b3oup2.ui"), \n',
             '         downloadButton("{prefix}b3oup2.pdf", "Download PDF"), \n',
             '         downloadButton("{prefix}b3oup2.png", "Download PNG"), br(), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}b3oup2.h", "PDF / PNG height:", width = "138px", \n',
             '                          min = 4, max = 20, value = 8, step = 0.5)), \n',
             '         div(style="display:inline-block", \n',
             '             numericInput("{prefix}b3oup2.w", "PDF / PNG width:", width = "138px", \n',
             '                          min = 4, max = 20, value = 10, step = 0.5)), \n',
             '         br(), h4("Cell numbers"), \n',
             '         dataTableOutput("{prefix}b3oup2.dt") \n',
             '       ) \n',
             '     ) \n',
             '   ),     # End of tab (2 space) \n',
             ' \n',
             ' ### Tab1.c1: violinplot / boxplot \n',
             ' tabPanel( \n',
             '    HTML("Violinplot / Boxplot"),  \n',
             '   h4("Cell information / gene expression violin plot / box plot"), \n',
             '   "In this tab, users can visualise the gene expression or continuous cell information ",  \n',
             '   "(e.g. Number of UMIs / module score) across groups of cells (e.g. libary / clusters).", \n',
             '   br(),br(), \n',
             '   fluidRow( \n',
             '     column( \n',
             '       3, style="border-right: 2px solid black", \n',
             '       selectInput("{prefix}c1inp1", "Cell information (X-axis):", \n',
             '                   choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                   selected = {prefix}def$grp1) %>%  \n',
             '         helper(type = "inline", size = "m", fade = TRUE, \n',
             '                title = "Cell information to group cells by",  \n',
             '                content = c("Select categorical cell information to group cells by",  \n',
             '                            "- Single cells are grouped by this categorical covariate",  \n',
             '                            "- Plotted as the X-axis of the violin plot / box plot")),  \n',
             '       selectInput("{prefix}c1inp2", "Cell Info / Gene name (Y-axis):", choices=NULL) %>%  \n',
             '         helper(type = "inline", size = "m", fade = TRUE, \n',
             '                title = "Cell Info / Gene to plot", \n',
             '                content = c("Select cell info / gene to plot on Y-axis", \n',
             '                            "- Can be continuous cell information (e.g. nUMIs / scores)", \n',
             '                            "- Can also be gene expression")), \n',
             '       radioButtons("{prefix}c1typ", "Plot type:", \n',
             '                    choices = c("violin", "boxplot"), \n',
             '                    selected = "violin", inline = TRUE), \n',
             '       checkboxInput("{prefix}c1pts", "Show data points", value = FALSE), \n',
             '{subst}       actionButton("{prefix}c1togL", "Toggle to subset cells"), \n',
             '{subst}       conditionalPanel( \n',
             '{subst}         condition = "input.{prefix}c1togL % 2 == 1", \n',
             '{subst}         selectInput("{prefix}c1sub1", "Cell information to subset:", \n',
             '{subst}                     choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                     selected = {prefix}def$grp1), \n',
             '{subst}         uiOutput("{prefix}c1sub1.ui"), \n',
             '{subst}         actionButton("{prefix}c1sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}         actionButton("{prefix}c1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}       ), br(), br(), \n',
             '       actionButton("{prefix}c1tog", "Toggle graphics controls"), \n',
             '       conditionalPanel( \n',
             '         condition = "input.{prefix}c1tog % 2 == 1", \n',
             '         sliderInput("{prefix}c1siz", "Data point size:",  \n',
             '                     min = 0, max = 4, value = {ptsiz}, step = 0.25),  \n',
             '         radioButtons("{prefix}c1psz", "Plot size:", \n',
             '                      choices = c("Small", "Medium", "Large"), \n',
             '                      selected = "Medium", inline = TRUE), \n',
             '         radioButtons("{prefix}c1fsz", "Font size:", \n',
             '                      choices = c("Small", "Medium", "Large"), \n',
             '                      selected = "Medium", inline = TRUE)) \n',
             '     ), # End of column (6 space) \n',
             '     column(9, uiOutput("{prefix}c1oup.ui"),  \n',
             '            downloadButton("{prefix}c1oup.pdf", "Download PDF"),  \n',
             '            downloadButton("{prefix}c1oup.png", "Download PNG"), br(), \n',
             '            div(style="display:inline-block", \n',
             '                numericInput("{prefix}c1oup.h", "PDF / PNG height:", width = "138px", \n',
             '                             min = 4, max = 20, value = 8, step = 0.5)), \n',
             '            div(style="display:inline-block", \n',
             '                numericInput("{prefix}c1oup.w", "PDF / PNG width:", width = "138px", \n',
             '                             min = 4, max = 20, value = 10, step = 0.5)) \n',
             '     )  # End of column (6 space) \n',
             '   )    # End of fluidRow (4 space) \n',
             ' ),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             '### Tab1.c2: Proportion plot \n',
             'tabPanel( \n',
             '  HTML("Proportion plot"), \n',
             '  h4("Proportion / cell numbers across different cell information"), \n',
             '  "In this tab, users can visualise the composition of single cells based on one discrete ", \n',
             '  "cell information across another discrete cell information. ",  \n',
             '  "Usage examples include the library or cellcycle composition across clusters.", \n',
             '  br(),br(), \n',
             '  fluidRow( \n',
             '    column( \n',
             '      3, style="border-right: 2px solid black", \n',
             '      selectInput("{prefix}c2inp1", "Cell information to plot (X-axis):", \n',
             '                  choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                  selected = {prefix}def$grp2) %>%  \n',
             '        helper(type = "inline", size = "m", fade = TRUE, \n',
             '               title = "Cell information to plot cells by",  \n',
             '               content = c("Select categorical cell information to plot cells by", \n',
             '                           "- Plotted as the X-axis of the proportion plot")), \n',
             '      selectInput("{prefix}c2inp2", "Cell information to group / colour by:", \n',
             '                  choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                  selected = {prefix}def$grp1) %>%  \n',
             '        helper(type = "inline", size = "m", fade = TRUE, \n',
             '               title = "Cell information to group / colour cells by", \n', 
             '               content = c("Select categorical cell information to group / colour cells by", \n',  
             '                           "- Proportion / cell numbers are shown in different colours")), \n',
             '      radioButtons("{prefix}c2typ", "Plot value:", \n',
             '                   choices = c("Proportion", "CellNumbers"), \n',
             '                   selected = "Proportion", inline = TRUE), \n',
             '      checkboxInput("{prefix}c2flp", "Flip X/Y", value = FALSE), \n',
             '{subst}      actionButton("{prefix}c2togL", "Toggle to subset cells"), \n',
             '{subst}      conditionalPanel( \n',
             '{subst}        condition = "input.{prefix}c2togL % 2 == 1", \n',
             '{subst}        selectInput("{prefix}c2sub1", "Cell information to subset:", \n',
             '{subst}                    choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                    selected = {prefix}def$grp1), \n',
             '{subst}        uiOutput("{prefix}c2sub1.ui"), \n',
             '{subst}        actionButton("{prefix}c2sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}        actionButton("{prefix}c2sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}      ), br(), br(), \n',
             '      actionButton("{prefix}c2tog", "Toggle graphics controls"), \n',
             '      conditionalPanel( \n',
             '        condition = "input.{prefix}c2tog % 2 == 1", \n',
             '        radioButtons("{prefix}c2psz", "Plot size:", \n',
             '                     choices = c("Small", "Medium", "Large"), \n',
             '                     selected = "Medium", inline = TRUE), \n',
             '        radioButtons("{prefix}c2fsz", "Font size:", \n',
             '                     choices = c("Small", "Medium", "Large"), \n',
             '                     selected = "Medium", inline = TRUE)) \n',
             '    ), # End of column (6 space) \n',
             '    column(9, uiOutput("{prefix}c2oup.ui"),  \n',
             '           downloadButton("{prefix}c2oup.pdf", "Download PDF"),  \n',
             '           downloadButton("{prefix}c2oup.png", "Download PNG"), br(), \n',
             '           div(style="display:inline-block", \n',
             '               numericInput("{prefix}c2oup.h", "PDF / PNG height:", width = "138px", \n',
             '                            min = 4, max = 20, value = 8, step = 0.5)), \n',
             '           div(style="display:inline-block", \n',
             '               numericInput("{prefix}c2oup.w", "PDF / PNG width:", width = "138px", \n',
             '                            min = 4, max = 20, value = 10, step = 0.5)) \n',
             '    )  # End of column (6 space) \n',
             '  )    # End of fluidRow (4 space) \n',
             '),     # End of tab (2 space) \n',
             ' \n',
             ' \n',
             ' \n',
             '  ### Tab1.d1: Multiple gene expr \n',
             '  tabPanel( \n',
             '    HTML("Bubbleplot / Heatmap"), \n',
             '    h4("Gene expression bubbleplot / heatmap"), \n',
             '    "In this tab, users can visualise the gene expression patterns of ", \n',
             '    "multiple genes grouped by categorical cell information (e.g. library / cluster).", br(), \n',
             '    "The normalised expression are averaged, log-transformed and then plotted.", \n',
             '    br(),br(), \n',
             '    fluidRow( \n',
             '      column( \n',
             '        3, style="border-right: 2px solid black", \n',
             '        textAreaInput("{prefix}d1inp", HTML("List of gene names <br /> \n',
             '                                          (Max 50 genes, separated <br /> \n',
             '                                           by , or ; or newline):"), \n',
             '                      height = "200px", \n',
             '                      value = paste0({prefix}def$genes, collapse = ", ")) %>% \n',
             '          helper(type = "inline", size = "m", fade = TRUE, \n',
             '                 title = "List of genes to plot on bubbleplot / heatmap", \n',
             '                 content = c("Input genes to plot", \n',
             '                             "- Maximum 50 genes (due to ploting space limitations)", \n',
             '                             "- Genes should be separated by comma, semicolon or newline")), \n',
             '        selectInput("{prefix}d1grp", "Group by:", \n',
             '                    choices = {prefix}conf[grp == TRUE]$UI, \n',
             '                    selected = {prefix}conf[grp == TRUE]$UI[1]) %>% \n',
             '          helper(type = "inline", size = "m", fade = TRUE, \n',
             '                 title = "Cell information to group cells by", \n',
             '                 content = c("Select categorical cell information to group cells by", \n',
             '                             "- Single cells are grouped by this categorical covariate", \n',
             '                             "- Plotted as the X-axis of the bubbleplot / heatmap")), \n',
             '        radioButtons("{prefix}d1plt", "Plot type:", \n',
             '                     choices = c("Bubbleplot", "Heatmap"), \n',
             '                     selected = "Bubbleplot", inline = TRUE), \n',
             '        checkboxInput("{prefix}d1scl", "Scale gene expression", value = TRUE), \n',
             '        checkboxInput("{prefix}d1row", "Cluster rows (genes)", value = TRUE), \n',
             '        checkboxInput("{prefix}d1col", "Cluster columns (samples)", value = FALSE), \n',
             '        br(), \n',
             '{subst}        actionButton("{prefix}d1togL", "Toggle to subset cells"), \n',
             '{subst}        conditionalPanel( \n',
             '{subst}          condition = "input.{prefix}d1togL % 2 == 1", \n',
             '{subst}          selectInput("{prefix}d1sub1", "Cell information to subset:", \n',
             '{subst}                      choices = {prefix}conf[grp == TRUE]$UI, \n',
             '{subst}                      selected = {prefix}def$grp1), \n',
             '{subst}          uiOutput("{prefix}d1sub1.ui"), \n',
             '{subst}          actionButton("{prefix}d1sub1all", "Select all groups", class = "btn btn-primary"), \n',
             '{subst}          actionButton("{prefix}d1sub1non", "Deselect all groups", class = "btn btn-primary") \n',
             '{subst}        ), br(), br(), \n',
             '        actionButton("{prefix}d1tog", "Toggle graphics controls"), \n',
             '        conditionalPanel( \n',
             '          condition = "input.{prefix}d1tog % 2 == 1", \n',
             '          radioButtons("{prefix}d1cols", "Colour scheme:", \n',
             '                       choices = c("White-Red", "Blue-Yellow-Red", \n',
             '                                   "Yellow-Green-Purple"), \n',
             '                       selected = "Blue-Yellow-Red"), \n',
             '          radioButtons("{prefix}d1psz", "Plot size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Medium", inline = TRUE), \n',
             '          radioButtons("{prefix}d1fsz", "Font size:", \n',
             '                       choices = c("Small", "Medium", "Large"), \n',
             '                       selected = "Medium", inline = TRUE)) \n',
             '      ), # End of column (6 space) \n',
             '      column(9, h4(htmlOutput("{prefix}d1oupTxt")), \n',
             '             uiOutput("{prefix}d1oup.ui"), \n',
             '             downloadButton("{prefix}d1oup.pdf", "Download PDF"), \n',
             '             downloadButton("{prefix}d1oup.png", "Download PNG"), br(), \n',
             '             div(style="display:inline-block", \n',
             '                 numericInput("{prefix}d1oup.h", "PDF / PNG height:", width = "138px", \n',
             '                              min = 4, max = 20, value = 10, step = 0.5)), \n',
             '             div(style="display:inline-block", \n',
             '                 numericInput("{prefix}d1oup.w", "PDF / PNG width:", width = "138px", \n',
             '                              min = 4, max = 20, value = 10, step = 0.5)) \n',
             '      )  # End of column (6 space) \n',
             '    )    # End of fluidRow (4 space) \n',
             '  )      # End of tab (2 space) \n',
             '  \n')
}

#' Append markers, all to the main block of ui.R
#'
#' @param prefix file prefix
#' @param markers.all TRUE/FALSE whether to include this tab
#' @rdname wrUImarkersAll
#' @export wrUImarkersAll
#'

wrUImarkersAll <- function(prefix, markers.all) {
  graphs <- ''
  if(markers.all == TRUE) {
    graphs <- graphs + glue::glue(
        '  , \n',
        '  \n',
        '  tabPanel( \n',
        '    HTML("Cluster Markers, All"), \n',
        '    h4("This page is meant to display associations between genetic markers and Seurat data clusters through an embedded spreadsheet."), \n',
        '    br(),br(), \n',
        '    fluidRow( \n',
        '      #selectInput("{prefix}m_meta_select", "Select meta:", choices=list({prefix}def$meta1, {prefix}def$meta1)), \n',
        '      #selectInput("{prefix}m_all_select", "Select cell:", choices=strsplit({prefix}conf[grp==TRUE][UI=={prefix}def$meta1]$fID, "\\\\|")[[1]]), # would you ever want to look at all? \n',
        '      uiOutput("{prefix}m_all_select.ui"), \n',
        '      column(12, \n',
        '        #DT::dataTableOutput("{prefix}m_all.ui") \n',
        '        uiOutput("{prefix}m_all.ui") \n',
        '      ) \n',
        '    ) \n',
        '  ) \n',
        .trim = FALSE
      )
  }
  return(graphs)
}

#' Append markers, top 20 to main block of ui.R
#'
#' @param prefix file prefix
#' @param markers.top20 TRUE/FALSE whether to include this tab
#' @rdname wrUImarkersTop20
#' @export wrUImarkersTop20
#'
wrUImarkersTop20 <- function(prefix, markers.top20) {
  graphs <- ''
  if(markers.top20 == TRUE) {
    graphs <- graphs + glue::glue(
        '  , \n',
        '  \n',
        '  tabPanel( \n',
        '    HTML("Cluster Markers, Top 20"), \n',
        '    h4("This page is meant to display associations between the top 20 genetic markers and Seurat data clusters through an embedded spreadsheet."), \n',
        '    br(),br(), \n',
        '    fluidRow( \n',
        '      column(12, \n',
        '        #DT::dataTableOutput("{prefix}m_t20") \n',
        '        uiOutput("{prefix}m_t20.ui") \n',
        '      ) \n',
        '    ) \n',
        '  ) \n',
        .trim = FALSE
      )
  }
  return(graphs)
}

#' Append differential gene expression to main block of ui.R
#'
#' @param prefix file prefix
#' @param markers.top20 TRUE/FALSE whether to include this tab
#' @rdname wrUIdeGenes
#' @export wrUIdeGenes
#'
wrUIdeGenes <- function(prefix, de.genes) {
  graphs <- ''
  if(de.genes == TRUE) {
    graphs <- graphs + glue::glue(
        '  , \n',
        '  \n',
        '  tabPanel( \n',
        '    HTML("Diff. Exp. Genes"), \n',
        '    h4("This page is meant to display associations between genetic markers and cell types tagged to their Seurat data cluster."), \n',
        '    br(),br(), \n',
        '    fluidRow( \n',
        '    #  conditionalPanel("{prefix}conf$DEs[1] == \"ERROR\"", \n',
        '    #                   HTML("<h5 style=\'color: red;\'>ERROR: a new \'{prefix}de_genes.rds\' was not created for this ShinyCellPlus instance!") \n',
        '    #  ), \n',
        '    #  conditionalPanel("{prefix}conf$DEs[1] != \"ERROR\"", \n',
        '    #                   selectInput("{prefix}de_genes_select", "Select differential expression:", choices=strsplit({prefix}conf$DEs[1], "\\\\|")[[1]]), # would you ever want to look at all? \n',
        '    #                   column(12, \n',
        '    #                     #DT::dataTableOutput("{prefix}de_genes.ui") \n',
        '    #                     dataTableOutput("{prefix}de_genes.ui") \n',
        '    #                   ) \n',
        '    #  ) \n',
        '      selectInput("{prefix}de_genes_select", "Select differential expression:", choices=strsplit({prefix}conf$DEs[1], "\\\\|")[[1]]), # would you ever want to look at all? \n',
        '      column(12, \n',
        '        #DT::dataTableOutput("{prefix}de_genes.ui") \n',
        '        #dataTableOutput("{prefix}de_genes.ui") \n',
        '        uiOutput("{prefix}de_genes.ui") \n',
        '      ) \n',
        '    ) \n',
        '  ) \n',
        .trim = FALSE
      )
  }
  return(graphs)
}

#' Append gene signature page to ui.R
#'
#' @param prefix file prefix
#' @param gene.ranks TRUE/FALSE whether to include this tab
#' @rdname wrUIgeneSig
#' @export wrUIgeneSig
#'

wrUIgeneSig <- function(prefix, gene.ranks) {
  graphs <- ''
  if(gene.ranks == TRUE) {
    graphs <- graphs + glue::glue(
      '  ,\n',
      '   \n',
      '  tabPanel( \n',
      '    HTML("Gene Signature"), \n',
      '    h4("View gene signature data through custom list of genes; uses AUCell for calculation."), \n',
      '    br(),br(), \n',
      '    fluidRow( \n',
      '      column(3, style="border-right: 2px solid black", \n',
      '        textAreaInput("gsig_list", "List of genes", value=paste0({prefix}def$genes, collapse = "\\n"), height="200px"), \n', #func to get correct genes??
      '        selectInput("gsig_group", "Group by:", choices={prefix}conf[grp==TRUE]$UI, selected={prefix}def$grp1), \n', #are these the correct groups?
      '        #actionButton("gsig_generate_plot", "Generate / Update plot", class = "btn-success"), \n',
      '        #br(),br(), \n',
      '        actionButton("gsig_subset_toggle", "Toggle to subset cells"), \n', #what kind of / how to get subsets? should be range of whatever group youre trying to subset
      '        conditionalPanel("input.gsig_subset_toggle % 2 == 1", \n',
      '                         selectInput("gsig_subset", "Cell information to subset:", choices={prefix}conf[grp==TRUE]$UI, selected={prefix}def$grp1), \n', #func to get correct groups
      '                         uiOutput("gsig_subset.ui"), \n',
      '                         actionButton("gsig_subset_sel_all", "Select all groups"), \n',
      '                         actionButton("gsig_subset_desel_all", "Deselect all groups") \n',
      '        ), \n',
      '        br(),br(), \n',
      '        actionButton("gsig_graphics_toggle", "Toggle graphics controls"), \n', #what kind of / how to get graphics controls?
      '        conditionalPanel("input.gsig_graphics_toggle % 2 == 1", \n',
      '                #radioButtons("gsig_graphics_plot_size", "Plot size:", choices = c("Small", "Medium", "Large"), inline=TRUE), \n',
      '                radioButtons("gsig_graphics_font_size", "Font size:", choices = c("Small", "Medium", "Large"), inline=TRUE), \n',
      '                sliderInput("gsig_graphics_point_size", "Point size:", min=0, max=4, value=0.60, step=0.10), \n',
      '                checkboxInput("gsig_graphics_cell_info", "Show cell info labels", value=TRUE), \n',
      '                radioButtons("gsig_graphics_color", "Color:", choices = c("White-Red", "Blue-Yellow-Red", "Yellow-Green-Purple"), selected="White-Red") \n',
      '        ) \n',
      '      ), \n',
      '      column(9, \n', 
      '        htmlOutput("gsig_list.html"), \n',
      '        plotOutput("gsig_plot", width="100%", height="600px"), \n',
      '        #htmlOutput("gsig_graphs.html"), \n',
      '        downloadButton("gsig_download.pdf", "Download PDF"), \n',
      '        downloadButton("gsig_download.png", "Download PNG"), \n',
      '        br(), \n',
      '        fillRow( \n',
      '          numericInput("gsig_download_height", "PDF / PNG height:", value=6, min=4, max=20, step=0.5, width="138px"), \n',
      '          numericInput("gsig_download_width", "PDF / PNG width:", value=14, min=4, max=20, step=0.5, width="140px"), \n',
      '          width="45%" \n',
      '        ) \n', 
      '      ) \n',
      '    ) \n',
      '  ) \n',
      '  \n',
      .trim = FALSE
    )
  }
  
  return(graphs)
}


#' Append gene differential gene expression volcano plot page to ui.R
#'
#' @param prefix file prefix
#' @param volc.plot TRUE/FALSE whether to include this tab
#' @rdname wrUIvolc
#' @export wrUIvolc
#'
wrUIvolc <- function(prefix, volc.plot) {
  graphs<-''
  if(volc.plot == TRUE) {
    graphs <- graphs + glue::glue(
      '  , \n',    
      '  \n',
      '  tabPanel( \n',
      '    HTML("Diff. Gene Exp., Volcano"), \n',
      '    h4("Generate volcano plots from differential gene expression analysis with Libra."), \n',
      '    br(),br(), \n',
      '    fluidRow( \n',
      '      column(3, style="border-right: 2px solid black", \n',
      '        #textAreaInput("volc_tags", "Genes to tag") \n',
      '        selectInput("volc_de_select", "Select differential expression:", choices=strsplit({prefix}conf$DEs[2], "\\\\|")[[1]]), # would you ever want to look at all? \n',
      '        actionButton("volc_subset_toggle", "Toggle to subset cells"), \n',
      '        conditionalPanel("input.volc_subset_toggle % 2 == 1", \n',
      '                         #selectInput("volc_subset", "Cell information to subset:", choices={prefix}conf[grp==TRUE]$UI, selected={prefix}def$grp1), \n',
      '                         #selectInput("volc_subset", "Cell information to subset:", choices={prefix}conf$DEs[3]), \n',
      '                         selectInput("volc_subset", "Cell information to subset:", choices=strsplit({prefix}conf$DEs[3], "\\\\|")[[1]]), \n',
      '                         uiOutput("volc_subset.ui"), \n',
      '                         actionButton("volc_subset_sel_all", "Select all groups"), \n',
      '                         actionButton("volc_subset_desel_all", "Deselect all groups") \n',
      '        ), \n',
      '        br(),br(), \n',
      '        textAreaInput("volc_gene_highlight", "Genes to Highlight"), \n',
      '        checkboxInput("volc_top10_toggle", "Toggle Top 10 Genes within Fold Change and p-Value Bounds", value=FALSE), \n',
      '        br(), \n',
      '        sliderInput("volc_fold_change", "Fold Change", 0, 2, 0.3, round=-1, step=0.1), # dynamic max? \n',
      '        textInput("volc_p_value", "Adjusted p-Value", placeholder="0.05"), \n',
      '        sliderInput("volc_xlim_up", "Upperbound range of fold change", 0, 10, 5, round=-1, step=0.1), \n',
      '        sliderInput("volc_xlim_down", "Lowerbound range of fold change", -10, 0, -5, round=-1, step=0.1), \n',
      '        sliderInput("volc_ylim", "Ceiling of p-value", 30, 600, 300, round=1, step=1), \n',
      '        textInput("volc_down_col", "Downregulated color:", placeholder="#00798c"), \n',
      '        textInput("volc_up_col", "Upregulated color:", placeholder="#d1495b"), \n',
      '        textInput("volc_not_sig_col", "Not significant color:", placeholder="grey82") \n',
      '      ), \n',
      '      column(9, \n',
      '        plotOutput("volc_plot", width="100%", height="600px"), \n',
      '        downloadButton("volc_download.pdf", "Download PDF"), \n',
      '        downloadButton("volc_download.png", "Download PNG"), \n',
      '        br(), \n',
      '        fillRow( \n',
      '          numericInput("volc_download_height", "PDF / PNG height:", value=6, min=4, max=20, step=0.5, width="138px"), \n',
      '          numericInput("volc_download_width", "PDF / PNG width:", value=14, min=4, max=20, step=0.5, width="140px"), \n',
      '          width="45%" \n',
      '        ) \n',
      '      ) \n',
      '    ) \n',
      '  ) \n',
      .trim = FALSE
    ) 
  }

  return(graphs)
}

#' Append ToppGene ontology page to ui.R
#'
#' @param prefix file prefix
#' @param gene.ont TRUE/FALSE whether to include this tab
#' @rdname wrUIgeneOnt
#' @export wrUIgeneOnt
#'
wrUIgeneOnt <- function(prefix, gene.ont) {
  graphs <-''
  if(gene.ont == TRUE) {
    graphs <- graphs + glue::glue(
       '  , \n',   
       '  \n',
       '  tabPanel( \n',
       '    HTML("ToppGene Ontology"), \n',
       '    h4("Plot ontology data retrieved from ToppGene database (would need to re-run ShinyCellPlus scripts to requery ToppGene).  NOTICE:  There will be "), \n',
       '    h5("NOTICE: When many clusters are being shown with Cluster Dotplots, it will be required to adjust the height and width of the plot in order \n',
       '       to view the graphs properly.  The same is required for saving the .png\'s or .pdf\'s.  "), \n',
       '    br(),br(), \n',
       '    fluidRow( \n',
       '      column(2, \n',
       '        selectInput("gont_de_select", "Select differential expression:", choices=strsplit({prefix}conf$DEs[4], "\\\\|")[[1]]), \n',
       '        uiOutput("gont_cat_select.ui"), \n',  
       '        selectInput("gont_plot_select", "Select plot method:", choices=c("Balloon plot", "Cluster dotplot")), \n',
       '        conditionalPanel("input.gont_plot_select == \'Balloon plot\'", \n',
       '                         sliderInput("gont_balloon_val", "Number of categories:", 1, 6, 1, step=1) \n',
       '        ), \n',
       '        conditionalPanel("input.gont_plot_select == \'Cluster dotplot\'", \n',
       '                         uiOutput("gont_cluster_vals.ui"), \n',
       '                         actionButton("gont_cluster_vals_sel_all", "Select all clusters"), \n',
       '                         actionButton("gont_cluster_vals_desel_all", "Deselect all clusters"), \n',
       '                         checkboxInput("gont_show_empty", "Show clusters with no category data") \n',
       '        ), \n',
       '        br(), \n',
       '        h5("Adjust these plot height and plot width values to large values (e.g. 1200 x 1200) to view large Cluster Dotplots"), \n',
       '        textInput("gont_plot_y", "Plot height: ", value = "1200", placeholder = "1200"), \n',
       '        textInput("gont_plot_x", "Plot width: ", value = "1200", placeholder = "1200"), \n',
       '        sliderInput("gont_text_size", "Text size: ", 1, 15, 6, step=1), \n',
       '        downloadButton("gont_download.pdf", "Download PDF"), \n',
       '        downloadButton("gont_download.png", "Download PNG"), \n',
       '        br(), \n',
       '        fillRow( \n',
       '          numericInput("gont_download_height", "PDF / PNG height:", value=35, min=4, max=300, step=0.5, width="138px"), \n',
       '          numericInput("gont_download_width", "PDF / PNG width:", value=35, min=4, max=300, step=0.5, width="140px"), \n',
       '          width="75%" \n',
       '        ), \n',
       '        br(),br(),br(),br(), \n', 
       '        h5("Notice: You may need to adjust the height and width for the file to be legible when saving Cluster Dotplots") \n',
       '      ), \n',
       '      column(10, \n',
       '        uiOutput("gont_plot.ui") \n',
       '      ) \n',
       '    ) \n',
       '  ) \n',
    )
  }

  return(graphs)
}

#' Write code for final portion of ui.R
#'
#' @param footnote shiny app footnote
#' 
#' @rdname wrUIend
#' @export wrUIend
#'
wrUIend <- function(footnote) {
  if(is.list(footnote)){
    f1 = ''; f2 = ''; f3 = ''; f4 = ''; f5 = ''; f6 = ''; f7 = ''; f8 = ''
    if(!is.null(footnote$author)){ f1 = paste0('"',footnote$author,' ",')}
    if(!is.null(footnote$title)){  f2 = paste0('"',footnote$title,' ",')}
    if(!is.null(footnote$journal)){f3 = paste0('em("',footnote$journal,' "),')}
    if(!is.null(footnote$volume)){ f4 = paste0('strong("',footnote$volume,', "),')}
    if(!is.null(footnote$page)){   f5 = paste0('"',footnote$page,' ",')}
    if(!is.null(footnote$year)){   f6 = paste0('"(',footnote$year,') ",')}
    if(!is.null(footnote$doi)){    f7 = paste0('"doi: ',footnote$doi,' ",')}
    if(!is.null(footnote$link)){   f8 = paste0('a("[Link]", href = "',footnote$link,'", target="_blank"),')}
    f0 = paste0('strong("Reference: "),', f1, f2, f3, f4, f5, f6, f7, f8, 
                'style = "font-size: 125%;"')
  } else {
    f0 = paste0('"', footnote[[1]], '", style = "font-size: 125%;"')
  }
  
  glue::glue('   \n',
             '   \n',
             'br(), \n',
             'p({f0}), \n',
             'p(em("This webpage was made using "), a("ShinyCellPlus", \n',
             '   href = "https://github.com/BioinformaticsMUSC/ShinyCellPlus", target="_blank"), \n',
             '  em(", a fork of "), a("ShinyCell", href = "https://github.com/SGDDNB/ShinyCell", target="_blank")), \n',
             'br(),br(),br(),br(),br() \n',
             '))) \n',
             ' \n',
             ' \n',
             ' \n',
             ' \n')
}

#' Write code for google-analytics.html
#'
#' @param gaID Google analytics tracking ID (e.g. "UA-123456789-0")
#' 
#' @rdname wrUIga
#' @export wrUIga
#'
wrUIga <- function(gaID) {
  glue::glue('<!-- Global site tag (gtag.js) - Google Analytics --> \n',
             '  <script async src="https://www.googletagmanager.com/gtag/js?id={gaID}"></script> \n',
             '  <script> \n',
             '  window.dataLayer = window.dataLayer || []; \n',
             'function gtag(){{dataLayer.push(arguments);}} \n',
             'gtag(\'js\', new Date()); \n',
             ' \n',
             'gtag(\'config\', \'{gaID}\'); \n',
             '</script> \n')
}
